/* 
	--------------------------------------------------------------------
	--------------------------------------------------------------------
	This file is part of SNAPdragon.

    SNAPdragon is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    SNAPdragon is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with SNAPdragon.  If not, see <http://www.gnu.org/licenses/>. 
    --------------------------------------------------------------------
    --------------------------------------------------------------------
-*/

BIBLIOGRAPHY = {};

BIBLIOGRAPHY["wu2002temperature"] = `Wu, Peng, Shu-ichi Nakano, and Naoki Sugimoto. "Temperature dependence of thermodynamic properties for DNA/DNA and RNA/DNA duplex formation." The FEBS Journal 269.12 (2002): 2821-2830.`;
BIBLIOGRAPHY["xia1998thermodynamic"] = `Xia, Tianbing, et al. "Thermodynamic parameters for an expanded nearest-neighbor model for formation of RNA duplexes with Watson− Crick base pairs." Biochemistry 37.42 (1998): 14719-14735.`;
BIBLIOGRAPHY["santalucia1998unified"] = `SantaLucia, John. "A unified view of polymer, dumbbell, and oligonucleotide DNA nearest-neighbor thermodynamics." Proceedings of the National Academy of Sciences 95.4 (1998): 1460-1465.`;
BIBLIOGRAPHY["traut1994physiological"] = `Traut, Thomas W. "Physiological concentrations of purines and pyrimidines." Molecular and cellular biochemistry 140.1 (1994): 1-22.`;
BIBLIOGRAPHY["zuker1981optimal"] = `Zuker, Michael, and Patrick Stiegler. "Optimal computer folding of large RNA sequences using thermodynamics and auxiliary information." Nucleic acids research 9.1 (1981): 133-148.`;


BIBLIOGRAPHY["maoileidigh2011unified"] = `Maoiléidigh, Dáibhid Ó., et al. "A unified model of transcription elongation: what have we learned from single-molecule experiments?." Biophysical journal 100.5 (2011): 1157-1166.`;
BIBLIOGRAPHY["tadigotla2006thermodynamic"] = `Tadigotla, Vasisht R., et al. "Thermodynamic and kinetic modeling of transcriptional pausing." Proceedings of the National Academy of Sciences of the United States of America 103.12 (2006): 4439-4444.`;

BIBLIOGRAPHY["komissarova1997transcriptional"] = `Komissarova, Natalia, and Mikhail Kashlev. "Transcriptional arrest: Escherichia coli RNA polymerase translocates backward, leaving the 3′ end of the RNA intact and extruded." Proceedings of the National Academy of Sciences 94.5 (1997): 1755-1760.`;
BIBLIOGRAPHY["gnatt2001structural"] = `Gnatt, Averell L., et al. "Structural basis of transcription: an RNA polymerase II elongation complex at 3.3 Å resolution." Science 292.5523 (2001): 1876-1882.`;
BIBLIOGRAPHY["nudler1997rna"] = `Nudler, Evgeny, et al. "The RNA–DNA hybrid maintains the register of transcription by preventing backtracking of RNA polymerase." Cell 89.1 (1997): 33-41.`;

BIBLIOGRAPHY["imashimizu2013direct"] = `Imashimizu, Masahiko, et al. "Direct assessment of transcription fidelity by high-resolution RNA sequencing." Nucleic acids research 41.19 (2013): 9090-9104.`;
BIBLIOGRAPHY["rosenberger1981estimate"] = `Rosenberger, R. F., and Gillian Foskett. "An estimate of the frequency of in vivo transcriptional errors at a nonsense codon in Escherichia coli." Molecular and General Genetics MGG 183.3 (1981): 561-563.`;
BIBLIOGRAPHY["shaw2002use"] = `Shaw, Randal J., Nicholas D. Bonawitz, and Daniel Reines. "Use of an in vivo reporter assay to test for transcriptional and translational fidelity in yeast." Journal of Biological Chemistry 277.27 (2002): 24420-24426.`;
BIBLIOGRAPHY["yaginuma2014diversity"] = `Yaginuma, Hideyuki, et al. "Diversity in ATP concentrations in a single bacterial cell population revealed by quantitative single-cell imaging." Scientific reports 4 (2014).`;
BIBLIOGRAPHY["adelman2002single"] = `Adelman, Karen, et al. "Single molecule analysis of RNA polymerase elongation reveals uniform kinetic behavior." Proceedings of the National Academy of Sciences 99.21 (2002): 13538-13543.`;
BIBLIOGRAPHY["tolic2004diversity"] = `Tolić-Nørrelykke, Simon F., et al. "Diversity in the rates of transcript elongation by single RNA polymerase molecules." Journal of Biological Chemistry 279.5 (2004): 3292-3299.`;
BIBLIOGRAPHY["depken2009origin"] = `Depken, Martin, Eric A. Galburt, and Stephan W. Grill. "The origin of short transcriptional pauses." Biophysical journal 96.6 (2009): 2189-2193.`;
BIBLIOGRAPHY["artsimovitch2000pausing"] = `Artsimovitch, Irina, and Robert Landick. "Pausing by bacterial RNA polymerase is mediated by mechanistically distinct classes of signals." Proceedings of the National Academy of Sciences 97.13 (2000): 7090-7095.`;
BIBLIOGRAPHY["imashimizu2013intrinsic"] = `Imashimizu, Masahiko, et al. "Intrinsic translocation barrier as an initial step in pausing by RNA polymerase II." Journal of molecular biology 425.4 (2013): 697-712.`;
BIBLIOGRAPHY["kuhner2007force"] = `Kühner, Ferdinand, et al. "Force-induced DNA slippage." Biophysical journal 92.7 (2007): 2491-2497.`;
BIBLIOGRAPHY["neher2004dynamics"] = `Neher, Richard A., and Ulrich Gerland. "Dynamics of force-induced DNA slippage." Physical review letters 93.19 (2004): 198102.`;


BIBLIOGRAPHY["XXX"] = ``;
BIBLIOGRAPHY["XXX"] = ``;







