<html>
<head>
<title>About Pauser</title>

<meta name="viewport" content="width=device-width, initial-scale=0.41, maximum-scale=1" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="../../src/Resources/phylo.css">
<script language='JavaScript' type='text/JavaScript' src="../../src/Resources/jquery-3.1.1.min.js"></script>
<script language='JavaScript' type='text/JavaScript' src="../../src/Resources/bibliography.js"></script>
<script language='JavaScript' type='text/JavaScript' src="../src/Controller/freeEnergy.js"></script>
<script language='JavaScript' type='text/JavaScript' src="../src/Controller/plots.js"></script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      processEscapes: true
    }
  });
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-106202043-1', 'auto');
  ga('send', 'pageview');

</script>


<script language='JavaScript' type='text/JavaScript'>


function init(){

	IS_MOBILE = false; 
	IS_MOBILE = (/(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|ipad|iris|kindle|Android|Silk|lge |maemo|midp|mmp|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows (ce|phone)|xda|xiino/i.test(navigator.userAgent) 
		 || /1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(navigator.userAgent.substr(0,4)));

	numReferences = 0;
	referenceNumbers = {};
	addCitations();
	printContents();


	 
    $('document').ready(function(){
            
            

            
                
                    
            // If the url is linked to a section then make sure the section is not collapsed
            // Sections given id in this format: RateBind_ParamHelp. This means the linkable section is in the Parameters section
            if(window.location.hash) {
                console.log(window.location.hash);
                
                var id = window.location.hash.substring(1);
                    
                    
                // Find the contents button for this element
                var contents_atag = $("#" + id + "_con");  
                contents_atag.click();
                console.log("Clicking", contents_atag);  


                $(window.location.hash).addClass("helpSelected");
                
                
                // Add back the anchor to ensure that the page is in the right place. If the webpage takes too long to load and
                // the anchored section is hidden, then the page won't anchor
                var hash = window.location.hash;
                window.location.hash = hash;
                window.location.hash = window.location.hash;


            }
                
            
            

        MathJax.Hub.Queue(["Typeset",MathJax.Hub]);
        
        
 });
            
	
	
	
}


function minimiseSection(element, setTo = null){

    var toHide = $(element).attr("id");
    toHide = toHide.substring(0, toHide.length-3); // Remove 'Btn' from the end
    toHide = toHide + "div";
    //console.log(toHide, $("#" + toHide));

    
    if(setTo == false || (setTo == null && $(element).attr("disp") == "false")){
        $(element).html("&plus;")
        $("#" + toHide).hide(setTo == null ? 300 : 0);
        $(element).attr("disp", true);
    }
    else{
        $(element).html("&minus;")
        $("#" + toHide).show(setTo == null ? 300 : 0);
        $(element).attr("disp", false);
    }

}

function openReferences(){

    
    minimiseSection("#References_Btn", true);

}

$(window).on('hashchange', function() {
    $(window.location.hash).addClass("helpSelected");
});
    
            
            
            
            
             
function printContents(){

            
    $("#contentsLoader").remove();
            
    // Find all elements with section or subsection classes
    var sects = $(".section, .subsection, .subsubsection");
    var sectNum = 0;
    var subsectNum = 0;
    var subsubsectNum = 0;
    var sectionID = "";
    for (var i = 0; i < sects.length; i ++){
            
        var ele = $(sects[i]);
        var isSect = ele.hasClass("section");
        var isSubsect = ele.hasClass("subsection");
            
            
        // Give it an id
        var HTML = (isSect || isSubsect) ? ele.html().trim() : ele.attr("sect").trim();
        var id = HTML.split(" ").join("_");
        var contents_ele_id = id + "_con";
        ele.attr("id", id);
            
        if (isSect) sectionName = id;

        // OnClick
        var onclick = "minimiseSection('#" + sectionName + "_Btn')"; // Click the button to hide/show the section
        var onclick_show = "minimiseSection('#" + sectionName + "_Btn', true)"; // Click the contents entry to show the section
            
        var a_tag = `<a id="` + contents_ele_id + `" href='#` + id + `' onClick="` + onclick_show + `" style='text-decoration:none; color:` + (isSect ? "#0B522D" : "black") + `'>` + HTML + `</a>`;
          
        // Add the html to the contents section
        if (isSect){
            sectNum++;
            
            
            $("#contents").append("<div style='font-size:18px; color:#0B522D'>" + sectNum + " " + a_tag + "</div>");
                        
            
            ele.before(`<span id='` + id + `_Btn' class='minimise' style="font-size:30px; padding: 5 5" disp="true"  onClick="` + onclick + `">&plus;</span>`);
            ele.parent().next("div").attr("id", id + "_div");
            
            subsectNum = 0;
            subsubsectNum = 0;
        }else if (isSubsect){
            subsectNum++;
            $("#contents").append("<div style='font-size:17px'>&nbsp;&nbsp;&nbsp;&nbsp;" + sectNum + "." + subsectNum + " " + a_tag + "</div>");
            subsubsectNum = 0;
        }
        else {
            subsubsectNum++;
            $("#contents").append("<div style='font-size:16px'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;" + sectNum + "." + subsectNum + "." + subsubsectNum + " " + a_tag + "</div>");
        }
        
        
            
    }
    



}
            


function addCitations(){



	var DOIspanTemplate = `
		<span style="background-color:#ebe9e7; padding: 2 4" title="DOItitle">
			<b style="color:#A5CF19">DOI:</b>
			<a style="text-decoration:none; color:#A5CF19" href = "DOIhref">view</a>
		</span>
		
	
	`;

	var citationTags = $("cite");
	for (var i = 0; i < citationTags.length; i ++){
	
		
		var vals = $(citationTags[i]).attr("val").split(",");
		if (vals.length == 0) continue;
		
		var citationHTML = "<sup class='refLink'>(";
		
		for (var j = 0; j < vals.length; j ++){
		
			// Add the reference to the bottom of the page if it does not already exist
			var val = vals[j].trim();
			if ($("#" + val).length == 0){
			
				numReferences++;
				referenceNumbers[val] = numReferences;
				var referenceString = BIBLIOGRAPHY[val]["name"];
				var DOIhtml = "";
				if (BIBLIOGRAPHY[val]["DOI"] != null && BIBLIOGRAPHY[val]["DOI"] != "XXXX"){
					DOIhtml = DOIspanTemplate.replace("DOIhref", "https://doi.org/" + BIBLIOGRAPHY[val]["DOI"]);
					DOIhtml = DOIhtml.replace("DOItitle", "https://doi.org/" + BIBLIOGRAPHY[val]["DOI"]);
					
				}
				
				$("#references").append('<div class="ref" id="' + val + '">' + numReferences + ') ' +  referenceString + DOIhtml + '</div><br>');

			}
			
			citationHTML +=  `<a onclick="openReferences();" class='refLink'  href="#` + val + `">` + referenceNumbers[val] + `</a>`;
			if (j < vals.length-1) citationHTML += ",";
			
		}
		
		
		// Create the link to the reference
		citationHTML += ")</sup>";
		$(citationTags[i]).append(citationHTML);
		

	
	}



};



</script>

</head>

<body onload="init()">
	<div id="screenContainer" style="position:relative">
		<div id="screen" style="opacity:1;">


				

			<div id="main" style="background-color:white">
				
				
				<div id="header" class="header" style="width:100%; ">
					<table style="margin: 0px auto; width:100%">
						

						
						<tr>
							<td style="height:30px; width: 30%; vertical-align:bottom; ">

							</td>


                         <td  style="text-align:center; font-size: 20px; vertical-align:top; font-family: 'Palatino'; width: 40%;"> 
                                <div style="width:100%; text-align:center;">

                                         <a href="../" style="text-decoration:none;  color:#414b4b" title="Launch Pauser">
                                            <div style="font-size: 108px; margin-bottom: -22px "> Pauser </div>
                                            <div style="font-size: 18px; "> Predictor of Transcriptional Pause Sites </div>
                                        </a>
                                        
                                    
                                    
                                </div>
                                        
                            </td>
                            
                            

                            <td style="height:30px; width: 30%; vertical-align:bottom; text-align:right;"> 



                              <a href="../../" style="text-decoration:none; color:white; ">
                                    <span class="noselect" style="padding:5 5; font-size: 18px; background-color:#A5CF19; padding: 8 8;" title="Go back home">Home</span> 
                              </a>&nbsp;
                              <a href="../" style="text-decoration:none; color:white; ">
                                <span class="noselect" style="padding:5 5; font-size: 18px; background-color:#A5CF19; padding: 8 8; color:white" title="Launch Pauser">Pauser</span> 
                               </a>&nbsp;
                                    <span class="noselect" style="padding:5 5; font-size: 18px; background-color:#0B522D; padding: 8 8;  color:white; " title="Introductory tutorials and other information about Pauser">About</span>
                         
                                    
                            </td>




						</tr>

						
					</table>
				</div>
			   
				
				

				<br>
				<div class="about-block" style="width:70%; margin:auto">
					<div class="about-block-inner">
						<div class="block-subtitle" style="font-size:45px"> Prediction of transcriptional pause sites </div> <br>
					
						<b>Disclaimer:</b> Pauser is currently in prerelease. It has not been peer-reviewed or published. <br><br>
                            
                            
                         Transcription elongation is carried out by RNA polymerase (RNAP). This involves RNAP translocating the DNA in the $3^\prime \rightarrow 5^\prime$ direction, while 
                         the RNA grows sequentially in the $5^\prime \rightarrow 3^\prime$ direction.
                         In prokaryotes and eukaryotes, this process occurs at a mean rate of 20-100 bp/s<cite val="ryals1982temperature,vogel1994rna,mason2005distinction"></cite>.
                         However, individual RNAP molecules proceed erractically along their template. Approximately once every 100 bp there exists a pause site<cite val="herbert2006sequence,larson2014pause"></cite>, 
                         at which RNAP can take seconds or minutes to transcribe. Transcriptional pausing plays a range of important biological roles in gene expression<cite val="winkler2002mrna,palangat1998transcriptional,de2003slow"></cite>.
                         
                            
                            <br><br>
                            
                        Pauser is a web-based program dedicated to predicting the locations of pause sites in gene sequences in a binary classification framework.
                            
                       
                            
                        The two predictors - <span style="color:#008cba">SimPol</span> and a <span style="color:#FFA500">Naive Bayes classifier</span> - were trained on the <i>E. coli</i> RNAP using published data<cite val="larson2014pause"></cite>.
                        
                            
                            
                            
				
					</div>
				</div>
                    
                    
                    
                       
                    
                <br>
                <div class="about-block" style="width:70%; margin:auto">
                    <div class="about-block-inner">
                        <div class="block-subtitle" style="font-size:45px">Contents</div> <br>
                    
                            
                            
                            <div style="border-style:solid; padding: 10; display:inline-block; min-width:200px; border-width: 1px; border-color:#A5CF19"  id="contents">
                                
                                <table  id="contentsLoader" title="Loading...">
                                <tr>
                                    <td>
                                        <div class="loader"></div> 
                                    </td>
                                    <td>
                                        Loading...
                                    </td>
                                </tr>
                            </table>
                                
                            </div>
                        
                
                    </div>
                </div>
                    
                    
                                
                <br><br><br><br>
                <div class="main-block">
                    <div class="main-block-inner">
                        <div class="block-title"> 
                                <span class="section">Getting started</span>
                            </div> 
                            <div style="display:none">          
                                <div class="about-block" style="width:90%; margin:auto">
                                    <div class="about-block-inner">
                                                
                                                    
                                                    
                                        <div class="about-block">
                                            <div class="about-block-inner">
                                                <div id="sequenceSelection" class="subsection block-subtitle">Information icons</div> <br>
                                                
                                                
                                                    Help icons <img title="Help" src="../../src/Images/help.png" style="vertical-align:middle; width:25px; height:25px;"> are scattered throughout the interface of Pauser. Pressing one of these buttons will redirect to
                                                    the relevant section on this page.
                                                
                                        
                                            </div>
                                        </div>
                                        
                                        
                                        
                                       
                                        <div class="about-block">
                                            <div class="about-block-inner">
                                                <div class="subsection block-subtitle">Uploading nucleotide sequences</div>  <br>
                                                    
                                                
                                                    One or more nucleotide sequences must be uploaded to begin Pauser, in .fasta format. Ensure that the sequences are positive sense and are specified $5^\prime \rightarrow 3^\prime$.
                                                    Optionally, the known locations of pause sites within the sequence may be included, where a pause site is defined as the length of the mRNA when the pause occurs.
                                                    This is done by appending the pause site indices onto the accession lines like so:
                                                
                                                
                                                    <div style="font-family:'Courier new'; margin:30px; padding: 5">
                                                    
                                                    
                                                        &gt;seq1|pausesites=(2,6,18-20)<br>
                                                        AGCGTTAGGCGATTCGGGAAATGCGATTG<br>
                                                        &gt;seq2<br>
                                                        ATTCGGGATCATCGAT
                                                    
                                                    
                                                    </div>
                                                
                                                
            
                                            </div>
                                        </div>
                                        
                                        
                                        
                                        
                                       
                                        <div class="about-block">
                                            <div class="about-block-inner">
                                                <div class="subsection block-subtitle">Uploading a SimPol session</div>  <br>
                                                    
                                                
                                                    Optionally, a session file may be generated using SimPol and uploaded to Pauser, in .xml format. This will determine the kinetic model to use for
                                                    predicting pause sites. The session file specifies the parameters and model settings, including their prior distributions, and the number of simulations
                                                    to perform per sequence. The default XML file can be viewed <a href="../pauser.xml" class="xml_link">here</a> and session files can be generated at 
                                                    <a style="color:#A5CF19" href="www.polymerase.nz/simpol">www.polymerase.nz/simpol</a>.
                                                    
                                                    
                                                   
                                                    
            
                                            </div>
                                        </div>
                                        
                                        
                                                
                                    </div>
                                </div>
                            </div>
                
                
                        </div>
                    </div>
                </div>
                
                
                
                
                
                   
                <br><br><br><br>
                <div class="main-block">
                    <div class="main-block-inner">
                        <div class="block-title"> 
                                <span class="section">Prediction of transcriptional pause sites</span>
                            </div> 
                            <div style="display:none">          
                                <div class="about-block" style="width:90%; margin:auto">
                                    <div class="about-block-inner">
                                                
                                                    
                                        Let $X = (x_1, x_2, \dotso , x_L)$ be a sequence with $L$ nucleotides. Each site within $X$ is classified as being a pause site $C_i = \mathcal{P}$ or as not
                                        being a pause site $C_i = \mathcal{N}$.
                                     
                                        Site $i$ is defined as a <b>pause site</b> if the average time to transcribe $i+1$ is significantly long, as specified by an adjustable threshold.
                                        
                                    
                                    
                                        <br><br>
                                        Upon beginning Pauser, each uploaded sequence is analysed by two classifiers - SimPol and a Naive Bayes classifier (NBC). 
                                        Each classifier calculates the <b>evidence</b> of any given site being a pause site. Interpetation of evidence differs between classifiers.
                                        The evidence threshold of each classifier can be adjusted independently. Low thresholds result in more sites classified as pause sites, and
                                        a high threshold results in fewer sites classified as pause sites.
                                    
                                                    
                                        <div class="about-block">
                                            <div class="about-block-inner">
                                                <div id="sequenceSelection" class="subsection block-subtitle">Classifiers</div> <br>
                                                
                                                
                                                   
                                                    <div class="panelOuter">
                                                        <div class="panelInner" >
                                                            <span class="subsubsection block-subsubtitle" sect="SimPol">SimPol</span> <br><br>
                                                    
                                                                SimPol is a framework for developing kinetic models of transcription elongation. These models are physically
                                                                informed by the thermodynamics of nucleic acids<cite val="wu2002temperature,xia1998thermodynamic,santalucia1998unified"></cite>.
                                                                While these models lack predictive power when it comes to transcriptional pausing, they are explanatory models and can guide the 
                                                                development of testable hypotheses. To change the default model, see  
                                                                    <a href="#Uploading_a_SimPol_session" onclick="minimiseSection('#Getting_started_Btn', true)" style="text-decoration:none; color:#A5CF19">Uploading a SimPol session</a>.
                                                                
                                                                <br><br>
                                                                <b>Evidence:</b> the evidence that site $i$ is a pause site is equal to the median time that the mRNA has length $i$, averaged across 
                                                                all simulations of the sequence.
                                                    
                                                    
                                                        </div> 
                                                    </div> 
                                            
                                            
                                            
                                            
                                                <div class="panelOuter">
                                                        <div class="panelInner" >
                                                            <span class="subsubsection block-subsubtitle" sect="Naive Bayes">Naive Bayes</span><br><br>
                                                    
                                                                            
                                                                A Naive Bayes classifier (NBC) is a simple statistical technique and is a sequence-based model. All nucleotides within a window of size $w_1 + w_2$ around
                                                                site $i$ are examined by the model, and this is used to classify $i$ into one of the two classes. The NBC assumes independence between sites. While NBC 
                                                                has significantly stronger predictive power than SimPol, it lacks explanatory power.
                                                    
                                                    
                                                              
                                                    
                                                    
                                                                <br><br>
                                                                <b>Evidence:</b> the evidence that site $i$ is a pause site is equal to the log posterior ratio:
                                                    
                                                                $$
                                                                    \log \frac{P(C_i = \mathcal{P}|X)}{P(C_i = \mathcal{N}|X)} = \log \frac
                                                                                                       {P(\mathcal{P}) \sum\limits_{j=i-w_1}^{i+w_2} P(x_j| C_i = \mathcal{P}) } 
                                                                                                       {P(\mathcal{N}) \sum\limits_{j=i-w_1}^{i+w_2} P(x_j| C_i = \mathcal{N})  }
                                                                $$
                                                                
                                                                where $P(\mathcal{P})$ and $P(\mathcal{N})$ are the prior probabilities of a site being a pause site / not a pause site (equal to 0.007 and 0.993 respectively).
                                                    
                                                    
                                                        </div> 
                                                    </div> 
                                            
                                                
                                        
                                            </div>
                                        </div>
                                    
                                    
                                    
                                    
                                    
                                        <div class="about-block">
                                            <div class="about-block-inner">
                                                <div id="sequenceSelection" class="subsection block-subtitle">Evaluating classifier adequacy</div> <br>
                                                
                                                    If the locations of pause sites are already known (see
                                                        <a href="#Uploading_nucleotide_sequences" onclick="minimiseSection('#Getting_started_Btn', true)" style="text-decoration:none; color:#A5CF19">Uploading nucleotide sequences</a>
                                            
                                                    ), then the adequacy of the classifiers can be quantified. The recall, precision, and accuracy of each sequence, and the average across all sequences
                                                    are given. Let $T_i = \mathcal{P}$ if site $i$ is a known pause site, and let $T_i = \mathcal{N}$ if $i$ is known to not be a pause site.
                                            
                                                    <br><br>
                                                    <b>Recall</b> is the probability of a true pause being classified as a pause: $P(C_i = \mathcal{P}|T_i  = \mathcal{P}) = \frac{P(C_i = \mathcal{P} \; \cap \; T_i  = \mathcal{P}}{T_i  = \mathcal{P}}$
                                            
                                                    
                                                    <br><br>
                                                    <b>Precision</b> is the probability that a site classified as a pause is actually a pause: $P(T_i = \mathcal{P}|C_i  = \mathcal{P}) = \frac{P(C_i = \mathcal{P} \; \cap \; T_i  = \mathcal{P}}{C_i  = \mathcal{P}}$
                                            
                                            
                                            
                                                    <br><br>
                                                    <b>Accuracy</b> is the overall proportion of correct classifications: $P(T_i = \mathcal{P} \; \cap \; C_i  = \mathcal{P}) + P(T_i = \mathcal{N} \; \cap \; C_i  = \mathcal{N})$
                                            
                                                
                                        
                                            </div>
                                        </div>
                                    
                                    
                                    
                                    
                                    
                                    
                                     <div class="about-block">
                                            <div class="about-block-inner">
                                                <div id="sequenceSelection" class="subsection block-subtitle">ROC curve</div> <br>
                                                
                                                    If the locations of pause sites are already known, then a ROC curve will be constructed. A ROC analysis accounts for all possible evidence thresholds
                                                    of a classifier and explores true and false positive rates simultaneously. The total adequacy of a classifier can be summarised by integrating the area
                                                    under the curve (AUC) of a ROC plot. The AUC lies in the range [0,1]. A perfect classifier will have an AUC of 1, while a classifier with an AUC of 0.5
                                                    or less is no better than a coin toss. See review by Fawcett<cite val="fawcett2006introduction"></cite> for an introduction to ROC analyses.
                                                
                                        
                                            </div>
                                        </div>
                                       
                                        
                                                
                                    </div>
                                </div>
                            </div>
                
                
                        </div>
                    </div>
                </div>
                
                
            
            
                      
                <br><br><br><br>
                <div class="main-block">
                    <div class="main-block-inner">
                        <div class="block-title"> 
                                <span class="section">Software information</span>
                            </div> 
                            <div style="display:none">          
                                <div class="about-block" style="width:90%; margin:auto">
                                    <div class="about-block-inner">
                                                
                                                    
                                                    
                                        <div class="about-block">
                                            <div class="about-block-inner">
                                                <div id="sequenceSelection" class="subsection block-subtitle">Running Pauser from the command line</div> <br>
                                                
                                                
                                                    
                                        The following instructions will work for Linux/Mac. 
                                    
                                       
                                        
                                        <div style="padding: 10 10; background-color:#ebe9e7" >
                                            <b> 1. Download Pauser </b> from <a href="https://github.com/jordandouglas/jordandouglas.github.io">here</a> and unzip. Open up the command terminal and navigate to this folder.
                                        </div> <br>
                                        
                                        <div style="padding: 10 10; background-color:#ebe9e7" >
                                            <b> 2. Compile the ViennaRNA suite module.</b> RNAfold<cite val="lorenz2011viennarna"></cite> is used when enabled in the SimPol .xml file. <br><br>
                                            
                                            
                                            <div style="margin-left: 100px;">
                                                    
                                                <span class="code">cd src/asm/ViennaRNA</span><br>
                                                <span class="code">bash build_vrna.sh </span><br>
                                            </div>
                                        </div> <br>
                                    
                                    
                                    
                                        <div style="padding: 10 10; background-color:#ebe9e7" >
                                                <b> 3. Compile Pauser. </b> ViennaRNA must be successfully compiled first.<br><br>
                                            
                                            
                                            <div style="margin-left: 100px;">
                                                    
                                                <span class="code">cd ../../../pauser/src/asm/</span><br>
                                                <span class="code">bash build_vrna.sh </span><br>
                                            </div>
                                        </div> <br>
                                        
                                        <div style="padding: 10 10; background-color:#ebe9e7" >
                                            <b> 4. Run Pauser.</b> <br><br>
                                                <div class="code" style="margin-left: 100px;">
                                                    
                                                    bash pauser.sh -fasta path/to/sequences.fasta -o path/to/output.csv
                                                </div> <br>
                                                where <span class="code">sequences.fasta</span> is the file containing sequences, and <span class="code">output.csv</span> the file to save the results to. Optional arguments:<br><br>
                                                <div style="margin-left: 100px;">
                                                    
                                                    <span class="code">-h</span>: View the command line argument specifications. <br><br>
                                                        
                                            
                                                    <span class="code">-xml session.xml</span>: Load the specified SimPol session in .xml format. If unspecified, will load the default session at pauser/pauser.xml. <br><br>
                                                   
                                                    <span class="code">-nbc params.txt</span>: Load the specified naive Bayes classifier parameters. If unspecified, will load the default session at pauser/NBC_pauser.txt. <br><br>
                                                   
                                                 <!--
                                                    <span class="code">-nthreads N</span>: Parallelise the simulations over N threads. Multithreading does not work in the web browser, only from the command line. Default: 1. <br><br>
                                                 -->
                                            
                                                </div>
                                        </div>
                                                
                                        
                                            </div>
                                        </div>
                                        
                                                
                                    </div>
                                </div>
                            </div>
                
                
                        </div>
                    </div>
                </div>
                
                
                             
    
                <br><br><br><br>
                <div class="main-block">
                    <div class="main-block-inner">
                        <div class="block-title"> 
                        
                        
                                <span class="section">References</span>
                            </div> 
                                <div style="display:none">
                                    <br><br>
                                    <div class="about-block" style="width:90%; margin:auto">
                                        <div class="about-block-inner"  id="references">
                                                
                                                
                                                
                                        </div>
                                    </div>
                                </div>
                            </div>
                
                
                        </div>
                    </div>
                </div>
            
                                
                    
                <br><br><br><br>
                
				
				

	
	
	
	<br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br>
	<br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> 
    <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br>
    <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> 
    <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br>
    <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> 

	
	
		<div class="header">
			Pauser | Centre for Computational Evolution & School of Biological Sciences, University of Auckland | 2019
		</div>
	


		</div>
	</div>


</div>
</body>
</html>
