<html>
<head>
<title>About SimPol</title>
<link rel="icon" href="../../src/Images/transcribe.png">
<meta name="viewport" content="width=device-width, initial-scale=0.41, maximum-scale=1" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="../../src/Resources/styles.css">
<script language='JavaScript' type='text/JavaScript' src="../../src/Resources/jquery-3.1.1.min.js"></script>
<script language='JavaScript' type='text/JavaScript' src="../../src/Resources/bibliography.js"></script>
<script language='JavaScript' type='text/JavaScript' src="../src/Controller/freeEnergy.js"></script>
<script language='JavaScript' type='text/JavaScript' src="../src/Controller/plots.js"></script>


<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      processEscapes: true
    }
  });
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-106202043-1', 'auto');
  ga('send', 'pageview');

</script>


<script language='JavaScript' type='text/JavaScript'>


function init(){

	IS_MOBILE = false; 
	IS_MOBILE = (/(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|ipad|iris|kindle|Android|Silk|lge |maemo|midp|mmp|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows (ce|phone)|xda|xiino/i.test(navigator.userAgent) 
		 || /1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(navigator.userAgent.substr(0,4)));
         
            
            
    // Contents
    printContents();
            
	setImageSizes();
	numReferences = 0;
	referenceNumbers = {};
	addCitations();
           
            
            
    // Draw the model diagram canvases
    drawModelDiagrams();
            
            
    // Parameter links
    populateParameterLinks(); 
            
            
    $('document').ready(function(){
            
            

            
            
                
    	// If the url is linked to a section then make sure the section is not collapsed
    	// Sections given id in this format: RateBind_ParamHelp. This means the linkable section is in the Parameters section
    	if(window.location.hash) {
    		console.log(window.location.hash);
            
            var id = window.location.hash.substring(1);
                
                
            // Find the contents button for this element
            var contents_atag = $("#" + id + "_con");  
            contents_atag.click();
            console.log("Clicking", contents_atag);  


    		$(window.location.hash).addClass("helpSelected");
    		
    		
    		// Add back the anchor to ensure that the page is in the right place. If the webpage takes too long to load and
    		// the anchored section is hidden, then the page won't anchor
    		var hash = window.location.hash;
    		window.location.hash = hash;
    		window.location.hash = window.location.hash;


    	}
            
            
            

        MathJax.Hub.Queue(["Typeset",MathJax.Hub]);
    	
    	
 });
            

            
            
            
            
            
	
}


function drawModelDiagrams(){

	var paramsResult = { leftGBase: 6, rightGBase: 15, leftMBase: 6, rightMBase: 15, NTPtoAdd: "A",
					 mRNAPosInActiveSite: -5, NTPbound: false, nbases: 0, mRNALength: "L", activated:true, rateOfBindingNextBase:0,
					 bulgePos: [0], bulgedBase: [-1], bulgeSize: [0], partOfBulgeID: [0], nextBaseToCopy: 16, terminated: false };
					 
	paramsResult["k -2,-1"] = 0;
	paramsResult["k -1,-2"] = 0;
	paramsResult["k -1,0"]  = 0;
	paramsResult["k 0,-1"]  = 0;
	paramsResult["k 0,+1"]  = 0;
	paramsResult["k +1,0"]  = 0;
	paramsResult["k +1,+2"] = 0;
	paramsResult["k +2,+1"] = 0;
	paramsResult["k +2,+3"] = 0;
	paramsResult["k +3,+2"] = 0;
	paramsResult["kN 0,+1"] = 0;
	paramsResult["kN +1,0"] = 0;
	paramsResult["k release0"] = 0;



	stateHoverEvents = [];
	var elongationModel = {id: "simpleBrownian", name: "Simple Brownian ratchet model", allowBacktracking: true, allowHypertranslocation: true, 
						allowInactivation:true, deactivateUponMisincorporation: false, 
						allowGeometricCatalysis: false, allowmRNAfolding:false, allowMisincorporation:false, useFourNTPconcentrations:false,
						NTPbindingNParams: 2, assumeBindingEquilibrium: false, currentTranslocationModel: "HIGI_barriers",
						assumeTranslocationEquilibrium: false};
										
				
	// Simple kinetic model
	drawModelDiagramCanvas_givenParams("modelDiagramCanvas", "kineticStateDescription", paramsResult, elongationModel);
	
   
}

            
function populateParameterLinks(){
            
            
            
    var paramToAnchorTag = {
        NTPconc:{anchor:"[NTP]", latex:"[NTP]"},
        ATPconc:{anchor:"[ATP]", latex:"[ATP]"},
        CTPconc:{anchor:"[CTP]", latex:"[CTP]"},
        GTPconc:{anchor:"[GTP]", latex:"[GTP]"},
        UTPconc:{anchor:"[UTP]", latex:"[UTP]"},
        FAssist:{anchor:"Force", latex:"$F_{assist}$"},
        haltPosition:{anchor:"Halt", latex:"Halt"},

        hybridLen:{anchor:"Hybrid", latex:"$h$"},
        bubbleLeft:{anchor:"Bubble_left", latex:"$\\beta_1$"},
        bubbleRight:{anchor:"Bubble_right", latex:"$\\beta_2$"},

        DGtaudag:{anchor:"Translocation_Gibbs_barrier", latex:"$\\Delta G_{\\tau}^{\\dagger}$"},
        DGtau1:{anchor:"Posttranslocated_Gibbs_energy", latex:"$\\Delta G_{\\tau 1}$"},
        barrierPos:{anchor:"Translocation_force_dependency", latex:"$\\delta_1$"},

        arrestTime:{anchor:"Arrest_timeout", latex:"Arrest time"},
        kCat:{anchor:"Rate_of_catalysis", latex:"$k_{cat}$"},
        Kdiss:{anchor:"NTP_dissociation_constant", latex:"$K_D$"},
        RateBind:{anchor:"Rate_of_NTP_binding", latex:"$k_{bind}$"},

        RateActivate:{anchor:"Rate_of_reactivation", latex:"k_A"},
        RateDeactivate:{anchor:"Rate_of_inactivation", latex:"k_U"},
        DGtaudagM:{anchor:"Gibbs_barrier_of_backtracking", latex:"$\\Delta G_{\\tau -}^\\dagger$"},
        DGtaudagP:{anchor:"Gibbs_barrier_of_hypertranslocation", latex:"$\\Delta G_{\\tau +}^\\dagger$"},

        RateCleave:{anchor:"Rate_of_cleavage", latex:"$k_{cleave}$"},
        CleavageLimit:{anchor:"Cleavage_limit", latex:"$\\lambda_{cleave}$"},
        rnaFoldDistance:{anchor:"Fold_distance", latex:"Fold distance"},

        proposalWidth:{anchor:"Proposal_width", latex:"Proposal width"}, 
            
            
        assumeBindingEquilibrium:{anchor:"Binding_equilibrium", latex:"$\\mathbb{BE}$"},
        assumeTranslocationEquilibrium:{anchor:"Translocation_equilibrium", latex:"$\\mathbb{TE}$"},
        allowmRNAfolding:{anchor:"Enable_RNA_folding", latex:"$\\mathbb{RB}$"}
            
            
    };
            
            
    
    var paramsTags = $("param");
            
     
    // Replace all <param value="paramID"/> tags with <a> tags that contain latex and a link to the appropriate parameter
    for (var i = 0; i < paramsTags.length; i ++){
            
        var ele = $(paramsTags[i]);  
        var paramID = ele.attr("value");
        console.log("paramID", paramID);
        var newTag = `<a onclick="minimiseSection('#Parameters_and_Settings_Btn', true)" class="param_link" href="#` + paramToAnchorTag[paramID].anchor + `">` + paramToAnchorTag[paramID].latex + `</a>`;
            
        ele.replaceWith(newTag);

            
    }
            
            
            
            
            
}            
            
            
function printContents(){

            
    $("#contentsLoader").remove();
            
    // Find all elements with section or subsection classes
    var sects = $(".section, .subsection, .subsubsection");
    var sectNum = 0;
    var subsectNum = 0;
    var subsubsectNum = 0;
    var sectionID = "";
    for (var i = 0; i < sects.length; i ++){
            
        var ele = $(sects[i]);
        var isSect = ele.hasClass("section");
        var isSubsect = ele.hasClass("subsection");
            
            
        // Give it an id
        var HTML = (isSect || isSubsect) ? ele.html().trim() : ele.attr("sect").trim();
        var id = HTML.split(" ").join("_");
        var contents_ele_id = id + "_con";
        ele.attr("id", id);
            
        if (isSect) sectionName = id;

        // OnClick
        var onclick = "minimiseSection('#" + sectionName + "_Btn')"; // Click the button to hide/show the section
        var onclick_show = "minimiseSection('#" + sectionName + "_Btn', true)"; // Click the contents entry to show the section
            
        var a_tag = `<a id="` + contents_ele_id + `" href='#` + id + `' onClick="` + onclick_show + `" style='text-decoration:none; color:` + (isSect ? "#008cba" : "black") + `'>` + HTML + `</a>`;
          
        // Add the html to the contents section
        if (isSect){
            sectNum++;
            
            
            $("#contents").append("<div style='font-size:18px; color:#008cba'>" + sectNum + " " + a_tag + "</div>");
                        
            
            ele.before(`<span id='` + id + `_Btn' class='minimise' style="font-size:30px; padding: 5 5" disp="true"  onClick="` + onclick + `">&plus;</span>`);
            ele.parent().next("div").attr("id", id + "_div");
            
            subsectNum = 0;
            subsubsectNum = 0;
        }else if (isSubsect){
            subsectNum++;
            $("#contents").append("<div style='font-size:17px'>&nbsp;&nbsp;&nbsp;&nbsp;" + sectNum + "." + subsectNum + " " + a_tag + "</div>");
            subsubsectNum = 0;
        }
        else {
            subsubsectNum++;
            $("#contents").append("<div style='font-size:16px'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;" + sectNum + "." + subsectNum + "." + subsubsectNum + " " + a_tag + "</div>");
        }
        
        
            
    }
    



}

function setImageSizes(){


									
	if(IS_MOBILE){

		$(".main-block").css("width", "90%");

		$("#polStructure").width("80%");
		$("#polStructure").css("height", "");
		$("#structureParams").width("80%");
		$("#structureParams").css("height", "");
		$("#breakingPairs").width("80%");
		$("#breakingPairs").css("height", "");
		$("#exponential").width("80%");
		$("#exponential").css("height", "");

	}



}


function minimiseSection(element, setTo = null){

	var toHide = $(element).attr("id");
	toHide = toHide.substring(0, toHide.length-3); // Remove 'Btn' from the end
    toHide = toHide + "div";
    //console.log(toHide, $("#" + toHide));

	
	if(setTo == false || (setTo == null && $(element).attr("disp") == "false")){
		$(element).html("&plus;")
		$("#" + toHide).hide(setTo == null ? 300 : 0);
		$(element).attr("disp", true);
	}
	else{
		$(element).html("&minus;")
		$("#" + toHide).show(setTo == null ? 300 : 0);
		$(element).attr("disp", false);
	}

}

function openReferences(){

	
	minimiseSection("#References_Btn", true);

}

$(window).on('hashchange', function() {
  	$(window.location.hash).addClass("helpSelected");
});


function addCitations(){



	var DOIspanTemplate = `
		<span style="background-color:#ebe9e7; padding: 2 4" title="DOItitle">
			<b style="color:#008cba">DOI:</b>
			<a style="text-decoration:none; color:#008cba" href = "DOIhref">view</a>
		</span>
		
	
	`;

	var citationTags = $("cite");
	for (var i = 0; i < citationTags.length; i ++){
	
		
		var vals = $(citationTags[i]).attr("val").split(",");
		if (vals.length == 0) continue;
		
		var citationHTML = "<sup class='refLink'>(";
		
		for (var j = 0; j < vals.length; j ++){
		
			// Add the reference to the bottom of the page if it does not already exist
			var val = vals[j].trim();
			if ($("#" + val).length == 0){
			
				numReferences++;
				referenceNumbers[val] = numReferences;
				var referenceString = BIBLIOGRAPHY[val]["name"];
				var DOIhtml = "";
				if (BIBLIOGRAPHY[val]["DOI"] != null && BIBLIOGRAPHY[val]["DOI"] != "XXXX"){
					DOIhtml = DOIspanTemplate.replace("DOIhref", "https://doi.org/" + BIBLIOGRAPHY[val]["DOI"]);
					DOIhtml = DOIhtml.replace("DOItitle", "https://doi.org/" + BIBLIOGRAPHY[val]["DOI"]);
					
				}
				
				$("#references").append('<div class="ref" id="' + val + '">' + numReferences + ') ' +  referenceString + DOIhtml + '</div><br>');

			}
			
			citationHTML +=  `<a onclick="openReferences();" class='refLink'  href="#` + val + `">` + referenceNumbers[val] + `</a>`;
			if (j < vals.length-1) citationHTML += ",";
			
		}
		
		
		// Create the link to the reference
		citationHTML += ")</sup>";
		$(citationTags[i]).append(citationHTML);
		

	
	}



};



</script>

</head>

<body onload="init()">
	<div id="screenContainer" style="position:relative">
		<div id="screen" style="opacity:1;">


				

			<div id="main" style="background-color:white">
				
				
				<div id="header" class="header" style="width:100%; ">
					<table style="margin: 0px auto; width:100%">
						

						
						<tr>
							<td style="height:30px; width: 30%; vertical-align:bottom; ">

							</td>

							<td  style="text-align:center; font-size: 20px; vertical-align:top; font-family: 'Palatino'; width: 40%;"> 
								<div style="width:100%; text-align:center;">

                                        <a href="../" style="text-decoration:none;  color:#414b4b" title="Launch SimPol">
                                            <div style="font-size: 90px; margin-bottom: -20px "> SimPol </div>
                                            <div style="font-size: 20px; "> Simulation of Nucleic Acid Polymerases </div>
                                        </a>
									
									
								</div>
										
							</td>
							
							

							<td style="height:30px; width: 30%; vertical-align:bottom; text-align:right;"> 


                              <a href="../../" style="text-decoration:none; color:white; ">
                                    <span class="noselect" style="padding:5 5; font-size: 18px; background-color:#008CBA; padding: 8 8;" title="Go back home">Home</span> 
                              </a>&nbsp;
                              <a href="../" style="text-decoration:none; color:white; ">
                                    <span class="noselect" style="padding:5 5; font-size: 18px; background-color:#008CBA; padding: 8 8;" title="Launch SimPol">SimPol</span> 
                              </a>&nbsp;
                                    <span class="noselect" style="padding:5 5; font-size: 18px; background-color:#b20000; padding: 8 8;  color:white; " title="Introductory tutorials and other information about SimPol">About</span>
                                    
                         
                                    
							</td>



						</tr>

						
					</table>
				</div>
			   
				
				

				<br>
				<div class="about-block" style="width:70%; margin:auto">
					<div class="about-block-inner">
						<div class="block-subtitle" style="font-size:45px"> An interactive graphical simulation of nucleic acid polymerases  </div> <br>
					
						<b>Disclaimer:</b> SimPol is currently in prerelease. It has not been peer-reviewed or published. <br><br>
                            
                            
                        
						SimPol is a free open-source web-based program aimed to help researchers, teachers and students visualise transcription at the level of a single molecule. 
						<u>Transcription</u> is an essential pathway for all living organisms. It involves the copying of a template, typically DNA, 
						into messenger RNA (mRNA) which is used to produce protein. Transcription is carried out by an enzyme called RNA polymerase. 
                        This webpage contains tutorials about SimPol and is hopefully friendly to all users. 
			

						
				
					</div>
				</div>
                    
                    
                    
                    
                    
                    
                    <br>
                <div class="about-block" style="width:70%; margin:auto">
                    <div class="about-block-inner">
                        <div class="block-subtitle" style="font-size:45px">Contents</div> <br>
                    
                            
                            
                            <div style="border-style:solid; padding: 10; display:inline-block; min-width:200px; border-width: 1px; border-color:#008cba"  id="contents">
                                
                                <table  id="contentsLoader" title="Loading...">
                                <tr>
                                    <td>
                                        <div class="loader"></div> 
                                    </td>
                                    <td>
                                        Loading...
                                    </td>
                                </tr>
                            </table>
                                
                            </div>
                        
                
                    </div>
                </div>
				


                <br><br><br>
				<div class="main-block" >
					<div class="main-block-inner">
						<div class="block-title"> 
                             <span class="section">Getting Started</span>
                        </div> 
						<div style="display:none">
                                
                                
                                
                                
                            <br><br>
                            <div class="about-block">
                                <div class="about-block-inner">
                                    <div id="sequenceSelection" class="subsection block-subtitle">Information icons</div> <br>
                                    
                                    
                                        Help icons <img title="Help" src="../../src/Images/help.png" style="vertical-align:middle; width:25px; height:25px;"> are scattered throughout the interface of SimPol. Pressing one of these buttons will redirect you to
                                        the relevant section on this page.
                                        <br><br>
                                        
                                         Conversely, there are direct links from this page to SimPol: &nbsp;&nbsp;&nbsp;
                                        <span class="xml_link" style="text-align:center">
                                               <b>Example session:</b> 
                                               <a >Open</a> | 
                                               <a >View XML</a>
                                        </span> 
                                    
                                        &nbsp;&nbsp;&nbsp; The XML files that specify the session parameters can be downloaded or the session can be directly opened.
                                    
                            
                                </div>
                            </div>
                                
                                
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div id="sequenceSelection" class="subsection block-subtitle"> Sequence selection </div> <br>
									
									
									<img src="../../src/TutorialImages/sequenceParams.png" style="width:350px; padding: 10 30;" align=left></img>
									
									<br><br><br><br><br>
									
                                        
                                    To enter a nucleic acid sequence, open up the &#9776; Parameters side-navigation menu, under "Select sequence" choose "Input sequence", enter a sequence and then press Submit.
                                    A range of example sequences have also been provided.
                                        
                   
									
									SimPol is designed primarily for transcription with DNA-dependent RNA-polymerases. Your choice of template-type (DNA or RNA) and nascent-type (DNA or RNA) will dictate
									which thermodynamic parameters are used - Turner's RNA/RNA parameters<cite val="xia1998thermodynamic"></cite>, Sugimoto's DNA/RNA parameters<cite val="wu2002temperature"></cite>, or SantaLucia's DNA/DNA parameters<cite val="santalucia1998unified"></cite>. <br><br>
								
							
								</div>
							</div>
				
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div class="subsection block-subtitle"> Polymerase structure </div> <br>
								
									Polymerases read the template strand (ie. the gene) in the 3&prime; &rarr; 5&prime; direction, and copy it to create the nascent strand (eg. mRNA) which grows in the 5&prime; &rarr; 3&prime; direction. 
                                    The transcription bubble consists of a hybrid between the nascent strand and the template strand within the polymerase, and a set of unpaired 
                                    template bases upstream and downstream of the polymerase. In the example below, the transcription bubble contains 2 + 9 + 1 = 12 bp.  <br><br>
									
									<img id="polStructure" src="../../src/TutorialImages/polStructure.png" style="height:200px; padding: 10 30;" align=left></img>
									<img id="structureParams" src="../../src/TutorialImages/structureParams.png" style="height:200px; padding: 10 30;" align=right></img>
									
									 <br><br><br><br><br><br><br><br><br><br><br><br><br>
									The size of the transcription bubble differs between polymerases.
									These three parameters may have an effect on the activity of the polymerase and can be changed from the &#9776; Parameters side-navigation menu. 
							
								</div>
							</div>
                                
                                
                                   
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div id="toolbarHelp" class="subsection block-subtitle"> The toolbar </div> <br>
									<table class="toolbar" cellpadding=3 style="position:absolute; vertical-align:middle">
										
										<tr>
											<td>
												<div class="toolbarIcon" title="Reinitiate the polymerase"> <img src="../../src/Images/reset.png" style=" vertical-align: middle; padding:0 0; width:35px; height:35px"> </div>
											</td>
											<td>
												<div class="toolbarIcon simulateRow" title="Perform N simulations stochastically"> <img src="../../src/Images/dice.png" style="vertical-align: middle; width:35px; height:35px;">   </div>
											</td>
										</tr>
										<tr>
											
											<td>
												<div class="toolbarIcon" title="Clear data from the cache" > 
													<img src="../../src/Images/close.png" style="padding:5 0; height:25px;">
												</div>
											</td>
											<td>
												<div class="toolbarIcon" title="Replicate/transcribe N bases"> <img id='trans'  src="../../src/Images/transcribe.png" style="vertical-align: middle; width:35px; height:35px;">  </div>
											
											</td>
										</tr>
										<tr>
											<td>
												<div id="foldBtnDiv" class="toolbarIcon"  title="Fold the mRNA (using the ViennaRNA suite)"> <img  src="../../src/Images/folding.png" style="vertical-align: middle; width:35px; height:35px;" >  </input> </div>
											</td>
											<td>
												<div id="stutterBtn" class="toolbarIcon" title="Stutter N times" > <img src="../../src/Images/stutter.png" style="vertical-align: middle; width:35px; height:35px;">  </div>
											
											</td>
										<tr>
											<td colspan=2>
												<div class="toolbarIcon" style="width:90px"> <span style="font-size: 16px; display:none; cursor: auto"></span> <span id="numSimSpan"> <b>N=</b><input type="number" name="nbasesToSimulate" id="nbasesToSimulate" value=30 style="vertical-align: middle; text-align:left; width: 60px; background-color: #008CBA; color:white"> </span> </div>
											</td>
										</tr>
										
										
										<tr>
											<td colspan=2>
												<div class="toolbarIcon">
													<select class="dropdown" title="Duration of each reaction" style="vertical-align: middle; text-align:right;">
														<option value="slow">Slow</option>
														<option value="medium">Medium</option>
														<option value="fast">Fast</option>
														<option value="hidden">Hidden</option>
													</select>
												</div>
											</td>
										</tr>
										
										
										<tr>
											<td>
												<div class="toolbarIcon"  title="Download the current session in xml format"> <img id='saveSessionID'   src="../../src/Images/downloadGreen.png" style="padding:5 0; height:25px;"> </div>
											</td>
											<td>
												<div class="toolbarIcon"  title="Upload a session in xml format" > 
													<img id='loadSessionID' class="rotate180" src="../../src/Images/downloadGreen.png" style="padding:5 0; height:25px;">
												</div>
											</td>
										</tr>
									</table>
								
									<div style="margin-left: 150px">
										<i>The toolbar contains some essential functions you may want to familiarise yourself with.</i> <br><br>
                                            
										<b>Reinitiate</b> <img  src="../../src/Images/reset.png" style="padding:0 0; width:20px; height:20px"> will reset the polymerase to its initial position. This will
										not affect plots or any other outputs. If you have added a background distribution behind any parameters then they will be resampled. <br><br>
										<b>Stop</b> <img src="../../src/Images/stop.png"  style="vertical-align:middle; width:20px; height:20px;"> will immediately halt any polymerase acitivies taking place.
										This will not affect any parameters, plots or other outputs.  <br><br>
										<b>N = </b><input type="number" value=30 style="vertical-align: middle; text-align:left; width: 60px; background-color: #008CBA; color:white; display:inline-block"> is a variable. The meaning of <b>N</b> depends on which
										tool is being used.  <br><br>
										<b>Simulate</b> <img src="../../src/Images/dice.png"  style="vertical-align:middle; width:25px; height:25px;"> begins the random simulation starting from the current position
										of the polymerase. During the simulation, reactions are
										randomly sampled according to their rate constants, which can be tweaked in the parameters menu. <b>N</b> trials are performed in the simulation. When the polymerase
										finishes copying the sequence (or prematurely terminates) the next trial begins. The duration of each reaction in the simulation is determined by <b>Speed</b>.
										The simulation can be stopped early using the Stop <img src="../../src/Images/stop.png"  style="vertical-align:middle; width:20px; height:20px;"> button.  <br><br>
										<b>Clear cache</b> <img src="../../src/Images/close.png"  style="vertical-align:middle; height:20px;"> opens up a dialog where you can delete any kinetic or sequence data which has been saved
										during the simulation. The kinetic data is displayed in plots so think of this button as resetting the plots.  <br><br>
										<b>Copy</b> <img src="../../src/Images/transcribe.png"  style="vertical-align:middle; width:25px; height:25px;"> transcribes/replicates/copies the next <b>N</b> bases. The duration of this process is determined by <b>Speed</b>.
										Copying can be stopped early using the Stop <img src="../../src/Images/stop.png"  style="vertical-align:middle; width:20px; height:20px;"> button. <br><br>
										<b>Stutter</b> <img src="../../src/Images/stutter.png"  style="vertical-align:middle; width:25px; height:25px;"> copies the current base <b>N</b> times, therefore adding an insertion of
										size <b>N</b> into the nascent strand. Insertions are added through slippage. The duration of this process is determined by <b>Speed</b>.
										Stuttering can be stopped early using the Stop </img src="../../src/Images/stop.png"  style="vertical-align:middle; width:20px; height:20px;"> button. <br><br>
										<b>Fold</b> <img src="../../src/Images/folding.png"  style="vertical-align:middle; width:25px; height:25px;"> folds the mRNA into its predicted secondary structure and renders it onto the nascent strand. This function will only work
										if the nascent strand is single-stranded RNA. Note that this function is only cosmetic and will not affect the polymerase's kinetics. If you want the mRNA secondary structure to affect reactions
										then make sure that "Enable mRNA folding" is enabled. This function is disabled on mobile devices. Folding is performed by the ViennaRNA suite <cite val="lorenz2011viennarna"></cite>.  <br><br>
										<b>Speed</b> <select class="dropdown" title="Duration of each reaction" style="vertical-align: middle; text-align:right; display:inline-block">
											<option value="slow">Slow</option>
											<option value="medium">Medium</option>
											<option value="fast">Fast</option>
											<option value="hidden">Hidden</option>
										</select>
	
										is a variable which controls the speed of the animation while Copying, Stuttering or Simulating. If the speed is set to <b>slow</b>,
										<b>medium</b> or <b>fast</b>, then the polymerase will animate each reaction at a slow, medium or fast speed, respectively. If the speed is <b>hidden</b>,
										then the polymerase is no longer displayed and will not animate at all. The plots will update periodically. <b>Hidden</b> is the fastest setting. <br><br>
										
										<b>Save</b> <img src="../../src/Images/downloadGreen.png"  style="vertical-align:middle; height:20px;"> downloads the current session in XML format. If the page is refreshed or if your web browser crashes you will 
										lose your current settings. If you do not want to lose your settings then you should save. <br><br>
										<b>Load</b> <img src="../../src/Images/downloadGreen.png" class="rotate180"  style="vertical-align:middle; height:20px;"> uploads a saved XML session file from your local storage.
										 <br><br>
									</div>
								</div>
							</div>
							</div>
					</div>
				</div>
				
                    
                    
                    
                    
                    
                    
                    
                    
                                
                <br><br><br><br>
                <div class="main-block">
                    <div class="main-block-inner">
                        <div class="block-title"> 
                                <span class="section">States and Reactions</span>
                            </div> 
                            <div style="display:none">          
                                <br><br>
                                <div class="about-block" style="width:90%; margin:auto">
                                    <div class="about-block-inner">
                                                
                                                    
                                        
                                        
                                        
                                        
                                       
                                        <div class="about-block">
                                            <div class="about-block-inner">
                                                <div class="subsection block-subtitle">Kinetic model of transcription</div> 
                                                    
                                                
                                                
                                                    Let $S(L,\tau)$ denote the current state, where $L$ is the length of the mRNA and $\tau$ is the position of the RNA polymerase (RNAP)
                                                    active site with respect to the $3^\prime$ mRNA. 
                                                    Under the Brownian ratchet model, transcription elongation is a three step cycle <cite val="abbondanzieri2005direct,bai2004sequence,maoileidigh2011unified,erie1992single"></cite>. 
                                                    First, RNAP translocates from the <b>pretranslocated</b> position $S(L,0)$ into the <b>posttranslocated</b> position $S(L,1)$. 
                                                    Second, the incoming nucleoside triphosphate (NTP) binds to the active site of RNAP.
                                                    Finally, the bound NTP is incorporated onto the $3^\prime$ mRNA, returning the RNAP into the pretranslocated position $S(L+1,0)$. <br><br>
                                                
                                                
                                                    However, RNAP does not always conform to this pathway. Every once in a while RNAP <b>backtracks</b> $S(L,\tau), \tau$ &lt; 0 whereby RNAP slides upstream
                                                    causing the $3^\prime$ mRNA to exit through the NTP entry channel <cite val="komissarova1997transcriptional,cheung2011structural"></cite>. This renders the enzyme catalytically inactive.
                                                    It has been hypothesised that an intermediate state (IS) acts as a precursor for backtracking from the pretranslocated state <cite val="artsimovitch2000pausing,imashimizu2013intrinsic,saba2019elemental"></cite>.
                                                    Alternatively, RNAP may <b>hypertranslocate</b> $S(L,\tau), \tau$ &gt; 1 whereby RNAP slides further downstream causing the shortening of the DNA/RNA hybrid <cite val="yarnell1999mechanism,larson2008applied"></cite>. 

                                                    
                                                    <br><br>
                                                
                                                    <div class="scrollbar" style="overflow-x:scroll; width:100%; position:relative">
                                                        <div style='background-color:adadad; padding: 10 10;  width: 1400px; margin:auto'>
                                                            <div style='background-color: ebe9e7; padding: 10 10; text-align:center; font-size:15; font-family:Arial; overflow-y:auto'>
                                                                <span style='font-size: 30px'> Kinetic diagram </span>
                                                                <div id="modelDiagramDiv" style="text-align:center">
                                                                    <canvas id="modelDiagramCanvas" class="noselect" style="padding:10 10;" width=1220 height=400></canvas>
                                                                </div>
                                                                
                                                                <div style="font-size:24; height:60px;">
                                                                    <div id="kineticStateDescription" style="display:block; text-align:left">  </div>
                                                                </div>
                                                                <div style="font-size:15"> Hover over a state read its description. </div>
                                                                
                                                            </div>
                                                        </div>
                                                    </div>
                                                
                                                    <br>
                                                    In SimPol, backtracking, hypertranslocation, and the IS can each be enabled or disabled from the &#9776; Parameters side-navigation menu.  The kinetic diagram of the current model can be viewed using
                                                <input type=button value='View model diagram' title="View the kinetic diagram of this model" class="operation" style="width:160px; cursor:auto">.
                                                    Pressing the simulate button  <img src="../../src/Images/dice.png" style="vertical-align: middle; width:25px; height:25px;"> simulates the system according to the
                                                    currently specified pathway. See 
                                                <a href="#Simulating" onclick="minimiseSection('#Simulating_Btn', true)" style="text-decoration:none; color:#008cba">Simulating</a>.
                                                    
                                                    
                                                    
                                                   
                                                    
            
                                            </div>
                                        </div>
                                        
                                        
                                        
                                        
                                        <br><br>
                                        <div class="about-block">
                                            <div  class="about-block-inner">
                                                <div class="subsection block-subtitle" > Reactions </div>
                                            
                                                
                                                The polymerase can be controlled manually using the Navigation panel at the top of the main page. 
                                                
                                                
                                                <br><br>
                                                <div class="panelOuter">
                                                        <div class="panelInner" >
                                                    <span class="block-subsubtitle" sect="Translocation" > Translocation: </span>
                                                    
                                                        There are two translocation reactions: Backwards and Forwards. These operations move the polymerase one position left and right respectively, with respect to the 
                                                    template and nascent strand. These operations involve reconfiguring the base-pairs within the hybrid, and within the gene if it is double-stranded. If the polymerase moves beyond the 3&prime; end of the nascent strand, 
                                                    then the nascent strand will be released (termination). 
                                                </div> 
                                                </div>  
                                                
                                                
                                                <div class="panelOuter">
                                                        <div class="panelInner" >
                                                    <span class="block-subsubtitle" sect="Nucleotide incorporation" > Nucleotide incorporation: </span>
                                                    
                                                        If the active site is free and next to the 3&prime; end of the nascent strand, then the polymerase may bind NTP.
                                                        Once bound, the polymerase may catalyse the reaction, or it may release the NTP back into the cell. Reverse-catalysis may also be performed 
                                                        the Decay reaction, although this reaction is often considered to be irreversible. The canonical pathway of transcription 
                                                        elongation is a three step cycle: 1) forward translocation, 2) NTP binding, 3) catalysis. This may also be achieved with the Copy 
                                                        <img src="../../src/Images/transcribe.png"  style="vertical-align:middle; width:25px; height:25px;"></img> function. 
                                                    </div> 
                                                </div> 
                                                
                                                
                                                
                                                
                                                <div class="panelOuter">
                                                        <div class="panelInner" >
                                                            <span class="block-subsubtitle" sect="Inactivation" > Inactivation: </span>
                                                            
                                                            
                                                            It is hypothesised that the catalytically inactive intermediate state, IS, acts as a precursor for backtracking. The Deactivate reaction causes the
                                                            polymerase to enter the IS from the pretranslocated state, thus enabling backtracking to occur. The Activate reaction is the reverse.
                   
                                                       </div> 
                                                </div> 
                                                    
                                                <div class="panelOuter">
                                                        <div class="panelInner" >
                                                    <span class="block-subsubtitle" sect="Cleavage"> Cleavage: </span>
                                                    
                                                        When a nucleic acid polymerase backtracks for too long <i>in vivo</i> it can delay the production of its product and potentially cause roadblocks for other polymerases.
                                                        Many organisms have evolved elongation factors which serve to rescue the polymerase from this state. One mechanism these factors may act is by cleaving the 3<sup>&prime;</sup> 
                                                        end of the nascent strand when it pokes out the front of the polymerase during backtracking. This is carried out by GreA and GreB for prokaryotic
                                                        RNA polymerases and S-II for eukaryotic RNA polymerase II <cite val="toulme2000grea"></cite> and can be performed in SimPol using the cleavage reaction.
                                                    </div> 
                                                </div> 
                                                
                            
                                                
                                                
                                                
                                                <div class="panelOuter">
                                                        <div class="panelInner" >
                                                    <span class="block-subsubtitle"  sect="Slippage"> Slippage: </span>
                                                    
                                                        Sometimes the nascent and template strand temporarily melt and a bulge forms in one of the strands. This can be performed by the
                                                    Form Bulge reaction.
                                                    When this happens, a bulge forms within the duplex. This can occur within either the template or nascent strand (although in SimPol it can currently only occur within the nascent strand). When a bulge forms,
                                                    the active site may be opened up, allowing one of the template bases to be copied twice. This would lead to an insertion. The bulge may 
                                                    diffuse left or right along the
                                                     hybrid. This process is favoured if the hybrid sequence is repetitive (eg. AAAAAAAAA/UUUUUUUUU). When positioned at either end of the hybrid, the bulge may be absorbed. This three step process (bulge formation, diffusion and absorption)
                                                     is based off the models proposed by Neher et al.<cite val="neher2004dynamics,kuhner2007force"></cite> and is believed to be the mechanism behind transcriptional slippage. Iterative slippage and elongation can lead to large insertions. This is sometimes called stuttering and
                                                    can also be achieved with the Stutter <img src="../../src/Images/stutter.png"  style="vertical-align:middle; width:25px; height:25px;"></img> function.
                                            
                                                    </div> 
                                                </div> 
                                                
                                            </div>
                                        </div>
                                                 
                                        
                                        
                                        
                                                
                                                
                                    </div>
                                </div>
                            </div>
                
                
                        </div>
                    </div>
                </div>
                                
                    
                <br><br><br><br>
				<div class="main-block" >
					<div class="main-block-inner">
						<div class="block-title"> 
                        
                         <span class="section">Plots</span>
                        </div> 
						<div style="display:none">
						
						
								<div class="about-block">
									<div class="about-block-inner">
										
										
										The output of the simulation can be visualised using various graphs, which are updated in real-time during the simulation.
										These plots are displayed in the Plots panel and in the Sitewise panel. The display of each plot can be configured with its respective settings 
									<img src="../../src/Images/gear.png"  style="vertical-align:middle; height:20px;"></img>
									button. The raw data (.tsv) or a high resolution image (.png) can be downloaded <img src="../../src/Images/download.png"  style="vertical-align:middle; height:20px;">. The different	
									plots available are detailed below.
							
									</div>
								</div>
							
							
                            
                            
                                <table class="about-block" cellpadding="10">
                                
                                
                                    <tr>
                                        <td colspan="3">
                                            <div class="subsection block-subtitle">Distance vs time </div> <br>
                                        </td>
                                    </tr>
                                    <tr>
                                        <td colspan="1">
                                            <img src="../../src/TutorialImages/distanceVsTime.png" style="vertical-align:middle; width:100%;">
                                        </td>
                                    
                                    
                                        <td colspan="2">
                                            
                                            This graph shows the progress made by the polymerase over time. The blue line represents the current simulation
                                            and the grey lines represent previous simulations. The x- and y-axis ranges can be set manually in the settings <img src="../../src/Images/gear.png"  style="vertical-align:middle; height:20px;"></img>.
                                        
                                        </td>
                                    
                                    </tr>
                                
                                
                                
                                
                                
                                    <tr>
                                        <td colspan="3">
                                            <br> <div class="subsection block-subtitle">Time histogram </div> <br>
                                        </td>
                                    </tr>
                                    <tr>
                                        <td colspan="1" style="width:33%">
                                            Depending on the settings <img src="../../src/Images/gear.png"  style="vertical-align:middle; height:20px;"></img>, this will either show the total time taken to copy the template (one value per trial) or the time in
                                            between successive catalysis events.
                                        </td>
                                    
                                    
                                        <td colspan="1" style="width:33%">
                                            <img src="../../src/TutorialImages/pauseHistogram.png" style="vertical-align:middle; width:100%;">
                                          
                                        </td>
                                    
                                    
                                         <td colspan="1" style="width:33%">
                                            <img src="../../src/TutorialImages/pauseHistogram2.png" style="vertical-align:middle; width:100%;">
                                          
                                        </td>
                                    
                                    </tr>
                                
                                
                                
                                
                                
                                
                                    <tr>
                                        <td colspan="3">
                                            <br> <div class="subsection block-subtitle">Velocity histogram </div> <br>
                                        </td>
                                    </tr>
                                    <tr>
                                        <td colspan="1">
                                            <img src="../../src/TutorialImages/velocityHistogram1s.png" style="vertical-align:middle; width:100%;">
                                            <br> <span style="font-size:0.8em"> $w=1$ s </span>
                                    
                                        </td>
                                    
                                    
                                        <td colspan="1">
                                            <img src="../../src/TutorialImages/velocityHistogram20s.png" style="vertical-align:middle; width:100%;">
                                            <br> <span style="font-size:0.8em"> $w=20$ s </span>
                                        </td>
                                    
                                    
                                         <td colspan="1">
                                            These plots are histograms of the <i>instantaneous</i> velocities during elongation, as opposed to <i> mean </i> velocity. The instantaneous velocity at time $t$ is defined as the derivative of distance over time
                                            $v(t) = \frac{dx}{dt}$. However since the distance versus time function is discontinuous, there is no derivative. Instead, the instantaneous velocities are calculated as the mean
                                            velocities within non-overlapping fixed-sized windows. The window size parameter $w$ can be changed in the settings  <img src="../../src/Images/gear.png"  style="vertical-align:middle; height:20px;"></img>.
                                            The plots on the left illustrate the effect this parameter can have. Larger window sizes tend to underrepresent negative velocities, which correspond to net backwards translocation within the time window.
                                            
                                          
                                        </td>
                                    
                                    </tr>
                                
                                
                                
                                
                                    <tr>
                                        <td colspan="3">
                                            <br> <div class="subsection block-subtitle">Parameter plot</div> <br>
                                        </td>
                                    </tr>
                                    <tr>
                                        <td colspan="1">
                                            Any parameter in the &#9776; Parameters menu can be plotted onto this graph through the settings menu <img src="../../src/Images/gear.png"  style="vertical-align:middle; height:20px;"></img>.
                                            The parameter's value will remain constant unless changed either (1) manually, or (2) by adding a prior distribution
                                            <img src="../../src/Images/distn.png" style="vertical-align:middle; width:38px; height:12px;"></img> 
                                            (see <a href="#Adding_a_prior_distribution_to_a_parameter" onclick="minimiseSection('#Simulating_Btn', true)" style="text-decoration:none; color:#008cba">Adding a prior distribution to a parameter</a>).
                                                     
                                            Parameters are sampled from their prior distribution at the beginning of each simulation. 
                                            Therefore one additional datapoint is made available to this plot at the end of each trial. 
                                            Parameters can be compared with each other and with various measurements computed at the end of each simulation (mean velocity, mean catalysis time, mean total transcription time, and the final length of the
                                            nascent strand). 
                                        </td>
                                    
                                    
                                        <td colspan="1">
                                            <img src="../../src/TutorialImages/parameterHeatmap1.png" style="vertical-align:middle; width:100%;">
                                          
                                        </td>
                                    
                                    
                                         <td colspan="1">
                                            <img src="../../src/TutorialImages/parameterHeatmap2.png" style="vertical-align:middle; width:100%;">
                                        </td>
                                    
                                    </tr>
                                    <tr>
                                        <td colspan="1">
                                            <img src="../../src/TutorialImages/parameterHeatmap3.png" style="vertical-align:middle; width:100%;">
                                             <br><br>
                                            <div class="xml_link" style="text-align:center">
                                                <b>Elongation heatmap session:</b> 
                                                <a href="../?s=elongationHeatmap.xml">Open</a> | 
                                                <a href="Examples/elongationHeatmap.xml">View XML</a>
                                            </div> 
                                        </td>
                                    
                                    
                                        <td colspan="1">
                                            <img src="../../src/TutorialImages/parameterHeatmap4.png" style="vertical-align:middle; width:100%;">
                                            <br><br>
                                            <div class="xml_link" style="text-align:center">
                                                <b>Incorporation heatmap session:</b> 
                                                <a href="../?s=incorporationHeatmap.xml">Open</a> | 
                                                <a href="Examples/incorporationHeatmap.xml">View XML</a>
                                            </div> 
                                        </td>
                                    
                                    
                                         <td colspan="1">
                                            Points can be coloured to represent a third variable, such as velocity. The third image
                                            shows the relationship between two parameters - <param value="kCat"/> and <param value="DGtaudag"/> -  required to achieve a velocity of 10-20 bp/s. As the rate of catalysis increases, 
                                            the Gibbs energy barrier of translocation must increase with it in order to achieve the same velocity. 
                                        
                                            <br><br>
                                            The forth images shows the relationship between <param value="Kdiss"/> and <param value="NTPconc"/>. As the dissociation constant of NTP binding increases, the concentration of NTP needed to achieve the same
                                            velocity increases.
                                        
                                        </td>
                                    
                                    </tr>
                                
                                
                                
                                
                                
                                    <tr>
                                        <td colspan="3">
                                            <br> <div class="subsection block-subtitle">MCMC trace</div> <br>
                                        </td>
                                    </tr>
                                    <tr>
                                        <td colspan="1">
                                            The trace plot is used to view the trace of an Markov chain Monte Carlo (MCMC) simulation 
                                            (see <a href="#MCMC_ABC" onclick="minimiseSection('#Simulating_Btn', true)" style="text-decoration:none; color:#008cba">MCMC ABC</a>).
                                             
                                            This is instrumental in interpreting chain convergence. The adjustable burn-in period
                                            is denoted by the red line. The variable displayed on the y-axis as well as the x- and y- axis limits can
                                            be adjusted in the settings <img src="../../src/Images/gear.png"  style="vertical-align:middle; height:20px;">.
                                            This plot is only accessible after MCMC has been initiated. Alternatively, the 
                                            posterior distribution may be downloaded a as a .log file and opening in 
                                            <a href="http://www.beast2.org/tracer-2/">Tracer</a><cite val="rambaut2015tracer"></cite>.
                                            
                                        </td>
                                    
                                    
                                        <td colspan="1">
                                            <img src="../../src/TutorialImages/tracePlot.png" style="vertical-align:middle; width:100%;">
                                          
                                        </td>
                                    
                                    
                                         <td colspan="1">
                                            <img src="../../src/TutorialImages/tracePlot2.png" style="vertical-align:middle; width:100%;">
                                        </td>
                                    
                                    </tr>
                                
                                
                                
                                
                                
                                    <tr>
                                        <td colspan="3">
                                            <br> <div class="subsection block-subtitle">Time per site</div> <br>
                                        </td>
                                    </tr>
                                    <tr>
                                        <td colspan="3">
                                            <img src="../../src/TutorialImages/timePerSite.png" style="vertical-align:middle; width:100%;">
                                            
                                        </td>
                                    </tr>
                                    <tr>
                                        <td colspan="3">
                                            The time per site plots report the amount (or proportion) of time spent at each site in the template. Three options for the y-axis measure
                                            are available in the settings <img src="../../src/Images/gear.png"  style="vertical-align:middle; height:20px;">.
                                            First, the "Time to catalysis" of site $i$ is the mean time required to incorporate an NTP at site $i$, where the timer starts upon the incorporation of site $i-1$.
                                            Second, the "Time spent at site" of site $i$ is the mean time that the RNAP active site is positioned on $i$. This can illuminate where the RNAP is spending its time during backtracking.
                                            Third, the "Time per transcript length" of site $i$ is the mean time that the nascent strand has length $i$.
                                            
                                        </td>
                                    </tr>
                                    
                                    
                                    
                                
                                
                                
                                </table>
                            
								
																										
						</div>
					</div>		
				</div>
				
				<br><br><br><br>
				<div class="main-block" >
					<div class="main-block-inner">
						<div class="block-title">
                            <span class="section">Parameters and Settings </span>
                        </div> 
						<div style="display:none">
				
		
								
								
								<div class="about-block">
									<div class="about-block-inner">
										
											
											<div style="background-color: #008CBA; width:100%">
											
												<div class="parameter-bg" style="font-size:18px">
													
													<div class="subsection parameter-title" style="font-size:30px">Experimental Conditions </div>  
													<div class="parameter-title" style="font-size:18px">The concentrations of NTP (or dNTP if the nascent strand is DNA), and the size of the assisting force applied to the system</div>  <br> 
														<div class="subsubsection paramHelp" sect="Number of NTPs">
															<table style="">
															<tr title="Do you want one concentration for all 4 NTPs or one concentration for each one?">
																<td style="text-align:right; width:150px">
														 			1 concentration  
														 		</td>
														 		<td style="" >
															 		<label class="switch">
																 		 <input type="checkbox"> </input>
																 		 <span class="slider round notboolean"></span>
																	</label> 
																	<span style="font-size:15px; vertical-align:middle" >4 concentrations</span>
														 		</td>
													 			<td style="width:75%">
																If this is set to <b>1 concentration</b> then all four NTPs have the same concentration [NTP]. If this is set to <b>4 concentrations</b>
																then the four NTP concentrations are set independently: [ATP], [CTP], [GTP], and [UTP]. Default: 1 concentration 
																</td>
															</tr>
														
															</table>
														</div><br>
														<div class="subsubsection paramHelp" sect="[NTP]">
															<b>[NTP]</b> - concentration of all four NTPs (or dNTPs). Default: 1000 &mu;M (or 20 &mu;M) 
														</div><br>
														<div class="subsubsection paramHelp" sect="[ATP]">
															<b>[ATP]</b> - concentration of ATP (or dATP). Default: 3152 &mu;M (or 24 &mu;M)<cite val="traut1994physiological"></cite>
														</div><br>
														<div class="subsubsection paramHelp" sect="[CTP]">
															<b>[CTP]</b> - concentration of CTP (or dCTP). Default: 278 &mu;M (or 29 &mu;M)<cite val="traut1994physiological"></cite>
														</div><br>
														<div class="subsubsection paramHelp" sect="[GTP]">
															<b>[GTP]</b> - concentration of GTP (or dGTP). Default: 468 &mu;M (or 5.2 &mu;M)<cite val="traut1994physiological"></cite>
														</div><br>
														<div class="subsubsection paramHelp" sect="[UTP]">
															<b>[UTP]</b> - concentration of UTP (or dTTP). Default: 567 &mu;M (or 37 &mu;M)<cite val="traut1994physiological"></cite>
														</div><br>
																			   
														<div class="subsubsection paramHelp" sect="Force">
															<b style="font-size:20px">$F_{assist}$</b> - the size of the force being applied to the polymerase during a single-molecule experiment. Positive forces are assisting loads and increase
																	the rate of transcription <cite val="abbondanzieri2005direct"></cite>. Negative forces are hindering loads and decrease the rate of transcription. In a natural cellular environment this external force
																	is not applicable. Default: 0 pN
														</div><br>	
                                            
                                            
                                            
                                                        <div class="subsubsection paramHelp" sect="Halt">
                                                            <b style="font-size:20px">Halt</b> - the position where the polymerase is halted prior to the start of the experiment. Change this value to start transcription further downstream.
                                                            Default: 14 nt
                                                        </div><br>  
									
												</div>
												<br><br>
												<div class="parameter-bg" style="font-size:18px">
													
													<div class="subsection parameter-title" style="font-size:30px">Structural Parameters </div>  
													<div class="parameter-title" style="font-size:18px">Basepairing configuration within the transcription bubble and polymerase.</div>  <br> 
													
														<div class="subsubsection paramHelp" sect="Hybrid">
															<b>Hybrid length (h)</b> - number of basepairs in the template/nascent hybrid within the polymerase. As translocation rates are determined from sequence composition, this value may affect
																			   the predicted location of pause sites. Default: 11 bp 
														</div><br>				   
														<div class="subsubsection paramHelp" sect="Bubble left">			   
															<b>Bubble size 3&prime; (&beta;<sub>1</sub>)</b> - number of unpaired template bases 3&prime; (left hand side) of the hybrid. This parameter is only applicable if the template is double stranded. Default: 1 bp
														</div><br>	
														<div class="subsubsection paramHelp" sect="Bubble right">
															<b>Bubble size 5&prime; (&beta;<sub>2</sub>)</b> - number of unpaired template bases 5&prime; (right hand side) of the hybrid. This parameter is only applicable if the template is double stranded. Default: 0 bp
														</div><br>	
									
												</div>
                                        
                                        
                                        
                                        
                                        
                                        
                                        <br><br>
                                                <div class="parameter-bg" style="font-size:18px">
                                                    
                                                    <div class="subsection parameter-title" style="font-size:30px">NTP Incorporation Parameters </div>  
                                                    <div class="parameter-title" style="font-size:18px">Kinetics of binding/releasing NTP, and catalysing polymerisation.</div>  <br> 
                                                        
                                            
                                            
                                                        <div class="subsubsection paramHelp" sect="Binding equilibrium">
                                                            <table style="">
                                                            <tr title="Assume that NTP binding is an equilibrium process?">
                                                                <td style="width:320px" >
                                                                    <label class="switch">
                                                                         <input type="checkbox"> </input>
                                                                         <span class="slider round"></span>
                                                                    </label> 
                                                                    <span style="font-size:15px; vertical-align:middle" >Assume NTP binding equilibrium, $\mathbb{BE}$</span>
                                                                </td>
                                                                <td>
                                                                    <div style="margin-left:50px">
                                                                        
                                                                        If enabled, NTP binding is assumed to occur on such a rapid timescale that the NTP binding partial equilibrium assumption is applied. This will change how $K_D$ and $k_{bind}$ are interpreted. Default: false.
                                                                    </div>
                                                                </td>
                                                            </tr>
                                                        
                                                            </table>
                                                        </div><br>
                                            
                                            
                                                        <div class="subsubsection paramHelp" sect="NTP dissociation constant">
                                                            <b style="font-size:20px">$K_D$</b> - equilibrium constant of NTP dissociation. If binding is assumed to be an equilibrium process, $K_D$ is the ratio between the rates of NTP binding and release $K_D = \frac{k_{release}}{k_{bind}}$. Whereas, in a kinetic binding model $K_D$ is used in conjunction with $k_{bind}$ to calculate the rate of releasing
                                                            bound NTP: $k_{release} = K_D k_{bind}$. Default: 1.8 &mu;M
                                                        </div><br>
                                                        <div class="subsubsection paramHelp" sect="Rate of catalysis"> 
                                                            <b style="font-size:20px">$k_{cat}$</b> - rate constant of incorporating bound NTP, thus extending the 3<sup>&prime;</sup> end of the nascent chain by one nucleotide. When translocation is fast,
                                                                    catalysis is the rate limiting step in transcription. Default: 25 s<sup>-1</sup>
                                                        </div><br>
                                                        
                                                        <div class="subsubsection paramHelp" sect="Rate of NTP binding">
                                                            <b style="font-size:20px">$k_{bind}$</b> - second order rate constant used to calculate rate constant of binding the complementary NTP, which is calculated as [NTP] k<sub>bind</sub>. If binding is assumed to be
                                                                    an equilibrium process, this parameter has no meaning and the textbox will disappear.
                                                                     Default: 0.5448 &mu;M<sup>-1</sup>.s<sup>-1</sup>  
                                                        </div><br>
                                                        
                                                        
                                                                               
                                    
                                                </div>
                                        
                                        
                                        
                                        
												<br><br>
												<div class="parameter-bg" style="font-size:18px">
													
													<div class="subsection parameter-title" style="font-size:30px">Translocation Parameters </div>
                                                    <div class="parameter-title" style="font-size:18px">Energetics of translocation reactions and force dependency.</div>  <br> 
                                            
                                                    <div class="subsubsection paramHelp" sect="Translocation equilibrium">
                                                            <table style>
                                                            <tr title="Assume that translocation between the pre and posttranslocated states is an equilibrium process?">
                                                                <td style="width:320px" >
                                                                    <label class="switch">
                                                                         <input type="checkbox"> </input>
                                                                         <span class="slider round"></span>
                                                                    </label> 
                                                                    <span style="font-size:15px; vertical-align:middle" >Assume translocation equilibrium, $\mathbb{TE}$</span>
                                                                </td>
                                                                <td>
                                                                    <div style="margin-left:50px">
                                                                        
                                                                        If enabled, translocation between the pretranslocated and posttranslocated positions is assumed to be at equilibrium. Backtracking and hypertranslocation are not affected. This will change how $\Delta G^\dagger_\tau$ and $\delta_1$ are interpreted. Default: false.
                                                                    </div>
                                                                </td>
                                                            </tr>
                                                        
                                                            </table>
                                                        </div><br>
                                            
													
														<div class="subsubsection paramHelp" sect="Translocation Gibbs barrier">
															<b style="font-size:20px">$\Delta G^\dagger_{\tau}$</b> - Gibbs energy barrier of translocation. To translocate forwards or backwards, several
															basepairs may need to be broken. The energy barrier which must be overcome in order to translocate is estimated as the energy needed to break these basepairs
															plus the value of this parameter. Increasing this value will make translocation slower. When translocation is assumed to reach at equilibrium, this parameter only
                                                            applies to backtracking and hypertranslocation. Default: 9.079 k<sub>B</sub>T <cite val="douglas2018bayesian"></cite>
														</div><br>	
														
														<div class="subsubsection paramHelp" sect="Posttranslocated Gibbs energy">
															<b style="font-size:20px">$\Delta G_{\tau 1}$</b> - Gibbs energy term added to that of the posttranslocated state. When this value is low,
															RNA polymerase spends more time in the posttranslocated state (compared with the pretranslocated state). Default: -2.007 k<sub>B</sub>T <cite val="douglas2018bayesian"></cite>
														</div> <br>	
														
														<div class="subsubsection paramHelp" sect="Translocation force dependency">
															<b style="font-size:20px">$\delta_1$</b> - position of the translocation transition state, compared with the two adjacent translocation positions. 
															This parameter lies between 0 and the width of a single basepair in DNA (3.4 &#x212b;<cite val="watson1953molecular"></cite>). This term only applies
															when translocation is kinetic and when an assisting force $F_{assist}$ is applied. Default: 2.838 &#x212b; <cite val="douglas2018bayesian"></cite>
														</div><br>	
														
									
												</div>
												
												<br><br>
												<div class="parameter-bg" style="font-size:18px">
													
													<div class="subsection parameter-title" style="font-size:30px">Pause Parameters </div>  
													<div class="parameter-title" style="font-size:18px">Kinetics of off-pathway translocation events.</div>  <br> 
                                            
                                            
                                            
                                                        <div class="subsubsection paramHelp" sect="Enable inactivation">
                                                            <table >
                                                            <tr title="Enable the polymerase to enter the intermediate state?">
                                                                <td style="width:320px" >
                                                                    <label class="switch">
                                                                         <input type="checkbox"> </input>
                                                                         <span class="slider round"></span>
                                                                    </label> 
                                                                    <span style="font-size:15px; vertical-align:middle" >Enable inactivation, $\mathbb{IS}$</span>
                                                                </td>
                                                                <td>
                                                                    <div style="margin-left:50px">
                                                                        
                                                                        If enabled, the intermediate state (IS) may be entered from the pretranslocated state. When enabled, this reaction is a necessary step for backtracking. Default: false.
                                                                    </div>
                                                                </td>
                                                            </tr>
                                                        
                                                            </table>
                                                        </div><br>
                                            
														<div class="subsubsection paramHelp" sect="Rate of inactivation">
															<b style="font-size:20px">$k_{U}$</b> - rate constant of polymerase entering the intermediate state. This state acts as a precursor for backtracking from the
                                                            pretranslocated state. Default: 0.1s<sup>-1</sup>
														</div><br>
														<div class="subsubsection paramHelp" sect="Rate of reactivation">
															<b style="font-size:20px">$k_{A}$</b> - rate constant of polymerase entering the pretranslocated state from the intermediate state. Default: 4s<sup>-1</sup>
														</div><br>
                                            
                                            
                                            
                                            
                                                        <div class="subsubsection paramHelp" sect="Enable backtracking">
                                                            <table >
                                                            <tr title="Enable the polymerase to backtrack during the simulation?">
                                                                <td style="width:320px" >
                                                                    <label class="switch">
                                                                         <input type="checkbox"> </input>
                                                                         <span class="slider round"></span>
                                                                    </label> 
                                                                    <span style="font-size:15px; vertical-align:middle" >Enable backtracking, $\mathbb{BT}$</span>
                                                                </td>
                                                                <td>
                                                                    <div style="margin-left:50px">
                                                                        
                                                                        If enabled, the polymerase may backtrack. If $\mathbb{IS}$ is also enabled, then entry into the intermediate state is necessary to begin backtracking. Default: false.
                                                                    </div>
                                                                </td>
                                                            </tr>
                                                        
                                                            </table>
                                                        </div><br>
                                            
                                            
                                            
                                                        <div class="subsubsection paramHelp" sect="Gibbs barrier of backtracking">
                                                            <b style="font-size:20px">$\Delta G^\dagger_{\tau -}$</b> - Gibbs energy barrier of backtracking. To translocate forwards or backwards, several
                                                            basepairs may need to be broken. The energy barrier which must be overcome in order to backtrack is estimated as the energy needed to break these basepairs
                                                            plus $\Delta G_\tau^\dagger$ plus the value of this parameter. Increasing this value will make backtracking slower. Default: 0 k<sub>B</sub>T 
                                                        </div><br>  
                                            
                                            
                                            
                                                        <div class="subsubsection paramHelp" sect="Enable hypertranslocation">
                                                            <table >
                                                            <tr title="Enable the polymerase to hypertranslocate during the simulation?">
                                                                <td style="width:320px" >
                                                                    <label class="switch">
                                                                         <input type="checkbox"> </input>
                                                                         <span class="slider round"></span>
                                                                    </label> 
                                                                    <span style="font-size:15px; vertical-align:middle" >Enable hypertranslocation, $\mathbb{HT}$</span>
                                                                </td>
                                                                <td>
                                                                    <div style="margin-left:50px">
                                                                        
                                                                        If enabled, the polymerase may hypertranslocate. This can lead to premature termination. Default: false.
                                                                    </div>
                                                                </td>
                                                            </tr>
                                                        
                                                            </table>
                                                        </div><br>
                                            
                                            
                                            
                                                         <div class="subsubsection paramHelp" sect="Gibbs barrier of hypertranslocation">
                                                            <b style="font-size:20px">$\Delta G^\dagger_{\tau +}$</b> - Gibbs energy barrier of hypertranslocation. To translocate forwards or backwards, several
                                                            basepairs may need to be broken. The energy barrier which must be overcome in order to hypertranslocate is estimated as the energy needed to break these basepairs
                                                            plus $\Delta G_\tau^\dagger$ plus the value of this parameter. Increasing this value will make hypertranslocation slower. Default: 0 k<sub>B</sub>T 
                                                        </div><br>  
                                            
                                            
														
												</div>
                                        
                                                 <br><br>
                                                <div class="parameter-bg" style="font-size:18px">
                                                        
                                                        <div class="subsection parameter-title" style="font-size:30px">RNA Folding Settings  </div>  
                                                        <div class="parameter-title" style="font-size:18px">Settings behind folding the nascent RNA strand.</div>  <br> 
                                                        
                                            
                                                            <div class="subsubsection paramHelp" sect="Fold distance">
                                                                <b style="font-size:20px">Fold distance</b> - the number of RNA nucleotides upstream and downstream from the polymerase which are sterically
                                                                prohibited from folding. Default: 8 nt <cite val="palangat1998transcriptional"></cite>
                                                            </div><br>  
                                            
                                            
                                            
                                                            <div class="subsubsection paramHelp" sect="Enable RNA folding">
                                                                <table >
                                                                <tr title="Enable the polymerase to hypertranslocate during the simulation?">
                                                                    <td style="width:320px" >
                                                                        <label class="switch">
                                                                             <input type="checkbox"> </input>
                                                                             <span class="slider round"></span>
                                                                        </label> 
                                                                        <span style="font-size:15px; vertical-align:middle" >Enable mRNA folding, $\mathbb{RB}$</span>
                                                                    </td>
                                                                    <td>
                                                                        <div style="margin-left:50px">
                                                                            
                                                                        If enabled, the nascent RNA strand will fold automatically during the simulation. The predicted energy barrier 
                                                                        of unfolding secondary structures in the mRNA may need to be overcome when the polymerase translocates backwards. When the nascent strand
                                                                        is not ssRNA, this feature is not applicable. Note that enabling this option will slow down the first trial of the simulation as the
                                                                        secondary structure of the RNA is computed (MFE algorithm <cite val="zuker1981optimal"></val></cite>) using the ViennaRNA suite <cite val="lorenz2011viennarna"></cite>. You can view the RNA secondary structure with the Fold 
                                                                        <img src="../../src/Images/folding.png"  style="vertical-align:middle; width:25px; height:25px;"></img> button. Default: false <br><br>
                                                                        </div>
                                                                    </td>
                                                                </tr>
                                                            
                                                                </table>
                                                            </div><br>
                      
                                                           
                                                    </div>
                                               
                                        
                                        
                                        
                                                <br><br>
                                                <div class="parameter-bg" style="font-size:18px">
                                                    
                                                    <div class="subsection parameter-title" style="font-size:30px">Cleavage Parameters </div>  
                                                    <div class="parameter-title" style="font-size:18px">Model for cleaving the nascent strand while backtracked.</div>  <br> 
                                                       <div class="subsubsection paramHelp" sect="Rate of cleavage" id="RateCleave_ParamHelp">
                                                            <b style="font-size:20px">$k_{cleave}$</b> - rate constant of cleaving the 3<sup>&prime;</sup> end of the nascent strand when the polymerase is backtracked. 
                                                            <i>In vivo</i> this reaction may be carried out by elongation factors, for example, GreA. Default: 0s <sup>-1</sup>
                                                        </div><br>
                                            
                                            
                                                     <div class="subsubsection paramHelp" sect="Cleavage limit" id="RateCleave_ParamHelp">
                                                            <b style="font-size:20px">$\lambda_{cleave}$</b> - the maximum number of backtracked nucleotides which can be cleaved. 
                                                            Set to 0 to have no upper limit. Default: 10 nt <cite val="lisica2016mechanisms"></cite>
                                                        </div><br>
                                                </div>
                                        
                                        
												<br><br>
												<div class="parameter-bg" style="font-size:18px">
													
													<div class="subsection parameter-title" style="font-size:30px">Advanced Model Settings</div>  
													<div class="parameter-title" style="font-size:18px">Other parameters relating to the algorithm and model.</div>  <br> 
													
                                            
                                                    <div class="subsubsection paramHelp" sect="Arrest timeout">
														<b>Arrest time </b> - the amount of time the polymerase may take to terminate before the simulation trial is aborted. When the timeout is
																reached, copying will prematurely terminate and the next trial will begin. Default: 0 s 
													</div><br>		
                                            
                                            
                                            
                                                    <div class="subsubsection paramHelp" sect="Proposal width">
                                                        <b>Proposal width </b> - scale width of operators during MCMC-ABC search. Higher values result in larger steps and smaller values give smaller steps.  Default: 1 
                                                    </div><br>      
                                            
                                                
													
												</div>
												<br><br>
											
										
											
											</div>
											
										
									</div>
								</div>
								
								
								
								<br><br>
								<div class="about-block">
									<div class="about-block-inner">
										<div id="priorDistribution_ParamHelp" class="subsection block-subtitle" > Adding a prior distribution to a parameter</div> <br>
											
											
											
											
											<br>
											If you are uncertain about the value of a parameter you can give it a background distribution <img src="../../src/Images/distn.png" style="vertical-align:middle; width:38px; height:12px;"></img> 
											to capture your uncertainty. The probability distributions supported are uniform, normal, lognormal, and exponential. The parameter is resampled at the beginning of each simulation. 
											All parameters are sampled independently of each other.
											<br><br>
											
											<!--
											<b>Example 1:</b> [ATP] is defaulted to Traut 1994’s<cite val="traut1994physiological"></cite> estimate of the mean ATP concentration (3152&mu;M) but one may wish to refine this value to reflect their own 
											experimental settings. Or one may recognise that cellular NTP concentrations are not constant, and they fluctuate throughout the cell cycle. In this case, it may be more accurate to 
											add a prior distribution behind [ATP]. Yaginuma et al. 2014 <cite val="yaginuma2014diversity"></cite> describe heterogeneity in [ATP] across E.coli cells with [ATP] following a positive-skewed modal distribution 
											(mean: 1540&mu;M, SD: 1220&mu;M). Using SimPol and some algebra, one could model [ATP] approximately with a lognormal distribution (where &mu; = 7.10, &sigma; = 0.698). Then the ATP 
											concentration would be resampled from this distribution at the beginning of each trial. <br><br>
											-->
											
									</div>
								</div>
									
								
											
								<br><br><br>
						</div>
					</div>		
				</div>
				<br><br><br><br>
				<div class="main-block" >
					<div class="main-block-inner">
						<div class="block-title">
                            <span class="section">Simulating</span>
                        </div> 
						<div style="display:none">
						
						
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div class="subsection block-subtitle" > Equilibrium and the Boltzmann distribution </div> <br>
								
										Gibb's free energy, $\Delta G$, describes how energetically favourable a state is. A system will spend more time in states which have low free energies. 
										Suppose we have a single molecule which can exist in a number of chemical states. When the system has reached equilibrium, the probability of the molecule being in state $s$ 
										at any given time is estimated with a Boltzmann distribution:
										
										<h3>
										$$ p(s) \propto e^{-\frac{\Delta G_s}{k_B T}}, $$
										</h3>
										
										where $\Delta G_s$ is the Gibbs energy of state $s$, $k_B$ is Boltzmann's constant ($1.3806 \times 10^{-23}$J.K$^{-1}$), and $T$ is the temperature in Kelvins. 
										Energy is expressed in units of $k_B T$, by taking the experimental energy (in units of J) and dividing by $k_B T$.   <br><br>
										
										For example, suppose a single molecule is in equilibrium between two states, $A$ and $B$. These two states have free energies of $\Delta G_A = -3k_B T$ and $\Delta G_B = -6k_B T$,
										respectively. We would therefore expect the molecule to be in state $A$ 4.7% of the time:
										
										
										<h3>
											$$p(A) = \frac{e^{-\frac{\Delta G_A}{k_B T}}}{e^{-\frac{\Delta G_A}{k_B T}} + e^{-\frac{\Delta G_B}{k_B T}}} = 
													 \frac{e^{-\frac{-3k_B T}{k_B T}}}{e^{-\frac{-3k_B T}{k_B T}} + e^{-\frac{-6k_B T}{k_B T}}}  = 
													 \frac{e^{3}}{e^{3} + e^{6}}  =
													 0.047.$$
										</h3>
										
			
										
										<br>
										The Boltzmann distribution assumes that the system has been fluctuating for long enough to achieve equilibrium. 
                                        However, since the competing processes of transcription 
                                        (see <a style="text-decoration:none; color:#008cba" href="#Reactions" onclick="minimiseSection('#States_and_Reactions_Btn', true)" style="text-decoration:none; color:black">Reactions</a>)
										occur on similar timescales, directly applying the Boltzmann distribution is not always a good assumption. 
                                        
							
								</div>
							</div>
							
							
							
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div class="subsection block-subtitle" > Kinetics and the Arrhenius equation </div> <br>
								
										
										Chemical kinetics is the study of rates of chemical processes, and how long reactions take to occur. A chemical reaction can be described with a rate constant $k$ (units s$^{-1}$). 
										$k$ describes the speed of the reaction - a higher rate means that more reactions are expected to occur in a given time. 
										
										
										
										Energy is required for a system to transit from one state to another. 
										The standard Gibbs energy of activation for a reaction is denoted by $\Delta G^\dagger$. The larger this barrier, the more time needed for random thermal collisions to provide
										enough energy to allow the system to make the transition. 
                                        Suppose a system is currently in state $s$, with Gibbs energy $\Delta G_s$. 
										The Arrhenius equation can estimate the rate constant of transiting from state $s$ to state $s^\prime$:
										
										
										<h3>
											$$ k_{s \rightarrow s^\prime} = A e^{-\frac{\Delta G^\dagger - \Delta G_s}{k_B T}}, $$
										</h3>
										
										where $A$ is a pre-exponential constant (units s$^{-1}$), and $\Delta G^\dagger$ is the absolute height of the Gibbs energy barrier between states $s$ and $s^\prime$. 
										Once the rate constants of all possible reactions are known, we can start the single-molecule simulation. 
										
										
										
							
								</div>
							</div>
							
							
							
							
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div class="subsection block-subtitle"> The Gillespie algorithm </div> <br>
								
                                    
                                    Transcription is simulated <img src="../../src/Images/dice.png"  style="vertical-align:middle; width:25px; height:25px;"></img> using the Gillespie algorithm<cite val="gillespie1977exact,lecca2013stochastic"></cite>. 
                                    
									A critical assumption of this method is that these reactions are Poisson processes; the amount of time until a reaction occurs is independent of all other reactions 
                                    and independent of the amount of time elapsed. 
									For example, suppose we were to continue flipping a coin until it lands on tails. The probability of getting tails on the next flip is not affected by how many times we have already flipped, and is not affected by how 
									many times someone else has flipped their coin. Similarly, whether a chemical reaction has been waiting 1 second or 1 hour for the reaction to occur, the probability of it happening right now is always the same.
									The amount of time until reaction $s \rightarrow s^\prime$ occurs follows the exponential distribution with rate $k$. <br><br>
									
									<img id="exponential" src="../../src/TutorialImages/exponential.png" style="width:60%; padding: 10 30;" align=left></img>
									
									If there are multiple chemical reactions competing with each other, then the reaction with the shortest reaction time will be selected. For example, suppose we have have a test-tube containing 
									a single molecule of compound $A$, which can spontaneously undergo two reactions: $A \rightarrow B$ or $A \rightarrow C$, with rate constants 0.001s<sup>-1</sup> and 0.002s<sup>-1</sup> respectively. 
									The test-tube will flash a different colour of light for each reaction. <br><br>
									
									
									The amount of time until either one of these reactions occurs is random, and exponentially distributed (see figures on the left). If we were to simulate this on a computer, we could 
									generate two exponential random variables, one from each distribution. For example, suppose we sampled time 769.14s for Exponential(0.001) and 339.81s for Exponential(0.002). Then 
									reaction $A \rightarrow C$ would be selected because it has the shorter reaction time. The probability of sampling reaction $A \rightarrow B$ is 
									
									
									<h3>
										$$ p(A \rightarrow B) = \frac{0.001}{0.001 + 0.002} = 0.333,$$
									</h3>
										
									
									
									and the probability of sampling reaction $A \rightarrow C$ is $0.667$. 
									

									
									
							
								</div>
							</div>
				
						
						
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div id="polymerisation_ModelHelp" class="subsection block-subtitle" > Kinetics of nucleotide incorporation</div> <br>
										
										Adding the next base onto the $3^\prime$ end of the nascent strand requires one translocation step 
                                        (see <a href="#Kinetics_of_translocation" onclick="minimiseSection('#Simulating_Btn', true)" style="text-decoration:none; color:#008cba">Kinetics of translocation</a>) 
                                    
                                     and at least three chemical steps<cite val="erie1992single"></cite>: <br>
										
										
										
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b>1. NTP binding.</b> A free nucleoside triphosphate (NTP) molecule enters the active site and binds to the active site and its complementary base in the template. 
										</div><br>
										
										
										
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b>2. Phosphodisester bond formation.</b> The $5^\prime$ phosphate of thenucleoside monophosphate (NMP) is linked to the $3^\prime$-OH of the nascent strand.
										</div><br>
		
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b>3. Pyrophosphate release.</b> The bound pyrophosphate is released back into solution.
										</div><br><br>	
										
										Any of these steps could be reversible.
										As there is not sufficent information to model all three steps, nucleotide incorporation has been compressed into a two step 
										process<cite val="bai2004sequence,abbondanzieri2005direct,maoileidigh2011unified"></cite>.
										
										
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b>1. NTP binding.</b> Step 1 from above. This is a reversible process. NTP binding/release is dictated by two parameters: <param value="RateBind"/> and <param value="Kdiss"/>. 
		
										</div><br>
										
										
										
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b>2. Catalysis.</b> Steps 2-3 from above. This is assumed to be an irreversible process. The rate of incorporating a bound nucleotide is <param value="kCat"/>.
										</div><br><br>	
										
										
										 
											<param value="RateBind"/> (units s$^{-1}$&mu;M$^{-1}$) is the second order rate constant of binding NTP.
											The rate of binding an NTP is equal to <param value="RateBind"/> $\times$ <param value="NTPconc"/>. Once the NTP has been bound, it may be released back into solution.
											The rate of release is calculated from the dissociation constant <param value="Kdiss"/> (units &mu;M). Rate of NTP release is equal to <param value="Kdiss"/> $\times$ <param value="RateBind"/>. 
											<br><br>
											
											
											When the NTP concentration is high, or if <param value="RateBind"/> is large, then NTP binding becomes so rapid that equilibrium between these two states is achieved. Under these conditions,
											the kinetics of NTP binding are no longer relevant. <param value="Kdiss"/> simply the ratio between the rates of binding and release, and can be 
                                            used to calculate the proportion of time that NTP is bound. For example, if <param value="Kdiss"/> = 35&mu;M and <param value="NTPconc"/> = 1000&mu;M,
											then the proportion of time that NTP is bound (while in this two-state system) is around 97%:
											
										
											
											<h3>
												$$ p(\text{bound}) = \frac{ \frac{\text{[NTP]}}{K_D} }{\frac{\text{[NTP]}}{K_D} + 1} = \frac{ \frac{1000\text{&mu;M}}{35 \text{&mu;M}} }{\frac{1000\text{&mu;M}}{35 \text{&mu;M}} + 1} = 0.966. $$
											</h3>
										
											<br>
											
									
                                             NTP binding/release can be approximated as an equilibrium process using <param value="assumeBindingEquilibrium"/>. 
                                            This assumption is equivalent to setting <param value="RateBind"/> to a high value, and is much faster to simulate.
											
											
											
								
							
								</div>
							</div>
						
						
						
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div id="translocation_ModelHelp" class="subsection block-subtitle" >Kinetics of translocation</div> <br>
								
                                    
                                        Let $\Delta G_{S(L, \tau)}$ be the Gibbs energy of state $S(L, \tau)$. This term is calculated using:
                                    
                                    
                                        <h3>
                                            $$
                                                \Delta G_{S(L, \tau)} = \Delta G_{S(L, \tau)}^{(bp)} + \begin{cases} \Delta G_{\tau 1} &\text{ if } \tau = 1 \\ 0 &\text { if } \tau \neq 1 \end{cases}
                                            $$
                                        </h3>
                                    
                                        where $\Delta G_{S(L, \tau)}^{(bp)}$ is computed from themodynamic parameters of nucleotide basepairing<cite val="xia1998thermodynamic,wu2002temperature,santalucia1998unified"></cite>
                                        and <param value="DGtau1"/> is the Gibbs energy added onto the posttranslocated position<cite val="douglas2018bayesian"></cite>. The basepairing energetic contribution
                                        of a state is computed by summing the Gibbs energy of the gene to that of the hybrid:
                                    
                                    
                                    
                                         <h3>
                                            $$
                                               \Delta G_{total}^{(bp)} = \Delta G_{gene}^{(bp)} + \Delta G_{hybrid}^{(bp)}.
                                            $$
                                        </h3>
                                    
                                    
                                    
										Basepairs may need to be broken in order for the polymerase to translocate forwards (right) or backwards (left) from its current position. For example, consider the state below. 
										<br><br>
										<img id="breakingPairs" src="../../src/TutorialImages/basepairbreaking.png" style="width:90%; display:block;  margin: 0 auto;" ></img>
										<br><br>
										<div style="line-height:35px">
											For the polymerase to translocate one step backwards, the 
										
											<span style="position:absolute; width:33px; "> 
												<img src="../../src/Images/Gg.png" style="width:15px; height:15px; position:absolute; left:0px"></img>
												<img src="../../src/Images/Cg.png" style="width:15px; height:15px; position:absolute; left:18px"></img>
												<img src="../../src/Images/Cm.png" style="width:15px; height:15px; top:18px; position:absolute; left:0px"></img>
												<img src="../../src/Images/Gm.png" style="width:15px; height:15px; top:18px; position:absolute; left:18px"></img>
											</span> 
											<span style="width:33;  display:inline-block;"></span> doublet on the left of the transcription bubble and the 
											<span style="position:absolute; width:33px; "> 
												<img src="../../src/Images/Cg.png" style="width:15px; height:15px; position:absolute; left:0px"></img>
												<img src="../../src/Images/Gg.png" style="width:15px; height:15px; position:absolute; left:18px"></img>
												<img src="../../src/Images/Gm.png" style="width:15px; height:15px; top:18px; position:absolute; left:0px"></img>
												<img src="../../src/Images/Cm.png" style="width:15px; height:15px; top:18px; position:absolute; left:18px"></img>
											</span> 
											<span style="width:33;  display:inline-block;"></span> doublet on the right of the hybrid must both be broken. For the polymerase
											to translocate one step forwards, the 
											<span style="position:absolute; width:33px; "> 
												<img src="../../src/Images/Tg.png" style="width:15px; height:15px; position:absolute; left:0px"></img>
												<img src="../../src/Images/Tg.png" style="width:15px; height:15px; position:absolute; left:18px"></img>
												<img src="../../src/Images/Am.png" style="width:15px; height:15px; top:18px; position:absolute; left:0px"></img>
												<img src="../../src/Images/Am.png" style="width:15px; height:15px; top:18px; position:absolute; left:18px"></img>
											</span> 
											<span style="width:33;  display:inline-block;"></span> doublet on the right of the transcription bubble and the 
											<span style="position:absolute; width:33px; "> 
												<img src="../../src/Images/Ag.png" style="width:15px; height:15px; position:absolute; left:0px"></img>
												<img src="../../src/Images/Tg.png" style="width:15px; height:15px; position:absolute; left:18px"></img>
												<img src="../../src/Images/Um.png" style="width:15px; height:15px; top:18px; position:absolute; left:0px"></img>
												<img src="../../src/Images/Am.png" style="width:15px; height:15px; top:18px; position:absolute; left:18px"></img>
											</span> 
											<span style="width:33;  display:inline-block;"></span> doublet on the left of the hybrid must both be broken. 
										</div>
										<br>
										
                                        Let $S(L,\tau)$ and $S(L,\tau+1)$ be two neighbouring states and let $T(L,\tau)$ be the translocation transition state between the two.
                                        The Gibbs energy of this transition state is calculated as:
                                    
                                    
                                         <h3>
                                            $$
                                                \Delta G_{T(L, \tau)} = \Delta G_{T(L, \tau)}^{(bp)} + \Delta G_\tau^\dagger + \begin{cases}    \Delta G_{\tau -}^\dagger &\text{ if } \tau \leq -1 \text{ (backtracking) }\\ 
                                                                                                                                                0 & \text { if } \tau = 0  \text{ (elongation) } \\
                                                                                                                                                \Delta G_{\tau +}^\dagger &\text { if } \tau \geq 1  \text{ (hypertranslocation) }
                                                                                                                                \end{cases}
                                            $$
                                        </h3>
                                    
                                    
                                        where $\Delta G_{T(L, \tau)}^{(bp)}$ is the Gibbs energy of basepairing in the transition state,
                                        <param value="DGtaudag"/> is the Gibbs energy barrier of translocation, 
                                        <param value="DGtaudagM"/> is the Gibbs energy barrier of backtracking, and
                                        <param value="DGtaudagP"/> is the Gibbs energy barrier of hypertranslocation. Estimating $\Delta G_{T(L, \tau)}^{(bp)}$ can be achieved through
                                        several methods (discussed below). After the translocation energy barrier has been estimated, the rates of translocating between
                                        $S(L,\tau)$ and $S(L,\tau+1)$ can be calculated using the Arrhenius equation:
                                    
											<h3>
											$$
												k_{S(L,\tau) \rightarrow S(L,\tau+1)} = k_{fwd}(S(L,\tau)) = A e^{-(\Delta G_{T(L, \tau)} - \Delta G_{S(L, \tau)})} \\
                                                k_{S(L,\tau+1) \rightarrow S(L,\tau)} = k_{bck}(S(L,\tau+1)) = A e^{-(\Delta G_{T(L, \tau)} - \Delta G_{S(L, \tau+1)})}
											$$
											</h3>
										
										
											using pre-exponential constant $A = 10^6$ s<sup>-1</sup>.	<br><br>
                                            SimPol has several methods for estimating $\Delta G_{T(L, \tau)}^{(bp)}$ (denoted by $\mathbb{TS}$):
                                    
                                    
                                    
                                    <br><br>
                                    <div class="panelOuter">
                                        <div class="panelInner" >
                                            <span class="block-subsubtitle" > i. Absolute model </span>
                                        
                                             Translocation rates are determined exclusively from the energies of the translocation states and not the transition states.
                                            This is the most simple model, where
                                            
                                            <h3>
                                            $$
                                                \Delta G_{T(L, \tau)}^{(bp)} = 0. 
                                            $$
                                            </h3>
                                            
                                
                                        </div> 
                                    </div>  
                                    
                                    
                                    
                                    
                                    <div class="panelOuter">
                                        <div class="panelInner" >
                                            <span class="block-subsubtitle" > ii. Midpoint model </span>
                                        
                                            The Gibbs energy of basepairing of the transition state between $S(L,\tau)$ and $S(L,\tau+1)$ is calculated as 
                                        
                                            <h3>
                                            $$
                                                \Delta G_{T(L, \tau)}^{(bp)} = \frac{\Delta G_{S(L,\tau)} + \Delta G_{S(L,\tau+1)}}{2}.
                                            $$
                                            </h3>
                                        
                                            
                                            
                                            This was initially described by Tadigotla et al.<cite val="maoileidigh2011unified,tadigotla2006thermodynamic"></cite>
                                            
                                            
                                
                                        </div> 
                                    </div>  
                                    
                                    
                                    
                                    
                                    <div class="panelOuter">
                                        <div class="panelInner" >
                                            <span class="block-subsubtitle" >  iii. Hybrid/bubble melting models </span>
                                        
                                             
                                             Let $G$ be the set of basepairs within the gene of state $S(L, \tau)$,
                                            and let $G^\prime$ be that of $S(L, \tau)$. Similarly, let $H$ and $H^\prime$ be the sets of basepairs within the hybrid of $S(L, \tau)$ and $S(L, \tau+1)$ respectively. 
                                            Then, there are four ways of directly evaluating $\Delta G_{T(L, \tau)}^{(bp)}$ using thermodynamic parameters:
                                            
                                            
                                            
                                            <h3>
                                            $$
                                                \Delta G_{T(L, \tau)}^{(bp)} = \begin{cases} 
                                                                                                \Delta G^{(bp)}(G \cap G^\prime, H \cap H^\prime) & \; \text{ if }  H_IG_I \text { model} \\
                                                                                                \Delta G^{(bp)}(G \cap G^\prime, H \cup H^\prime) & \; \text{ if }  H_IG_U \text { model} \\
                                                                                                \Delta G^{(bp)}(G \cup G^\prime, H \cap H^\prime) & \; \text{ if }  H_UG_I \text { model} \\ 
                                                                                                \Delta G^{(bp)}(G \cup G^\prime, H \cup H^\prime) & \; \text{ if }  H_UG_U \text { model}.
                                                                                \end{cases}
                                            $$
                                            </h3>
                                        
                                            
                                            $\Delta G^{(bp)}(x,y)$ denotes the Gibbs energy of a state comprised of a gene containing basepairs in set $x$ and a hybrid containing basepairs in set $y$. 
                                            These four models describe physical mechanisms of translocation, each describing the ordering of basepairing melting and forming required to translocate.
 
                                
                                        </div> 
                                    </div>  
                                    <br><br>
                                    
                                
                                    
                                    
      
                                    
									
									In optical trapping experiments, beads are tethered to single molecules of RNA polymerase and optical tweezers are used to exert a force (units pN) upon the elongation complex. 
									Positive forces  <param value="FAssist"/> $= F > 0$ act in the same direction as transcription (an assisting load) while negative forces $F < 0$ act in the opposite direction 
									(a hindering load). Increasing the assisting force serves to decrease the Gibbs energy barrier of forward translocation and increase the barrier of backwards translocation. This 
									process is modelled by adding an additional term<cite val="depken2009origin"></cite> to the Arrhenius equation:
										<h3>
											
                                            
                                            <h3>
                                                $$
                                                    k_{fwd}(S(L,\tau)) = A e^{-(\Delta G_{T(L, \tau)} - \Delta G_{S(L, \tau)} - F(\delta_1))} \\
                                                    k_{bck}(S(L,\tau)) = A e^{-(\Delta G_{T(L, \tau-1)} - \Delta G_{S(L, \tau)} - F(\delta - \delta_1))}
                                                $$
                                            </h3>
                                            
										</h3>
									where $\delta$ is the distance between consecutive basepairs ($\delta = 3.4$ &#8491; in dsDNA<cite val="watson1953molecular"></cite>), and 
                                    0 &lt; <param value="barrierPos"/> &lt; $\delta$ is the position of the transition state between the two neighbouring states.
                                        
                                        
                                        
                                                                                   
                                            
                                    <br><br>
                                    Translocation between the pretranslocated $\tau=0$ and posttranslocated $\tau=1$ positions can alternatively be modelled with an equilibrium approximation
                                    <param value="assumeTranslocationEquilibrium"/>. This assumption asserts that <param value="DGtaudag"/> is so small that translocation
                                    occurs on a timescale similar to NTP binding or incorporation. 
							
								</div>
							</div>
						
					
						</div>
					</div>
				</div>
				
				<br><br><br><br>
				<div class="main-block">
					<div class="main-block-inner">
                        <div class="block-title"> 
                            <span class="section">Approximate Bayesian computation</span> 
                        </div> 
                        <div style="display:none">
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
								<div class="subsection block-subtitle" >Bayesian inference</div> <br>
									
                                    
                                    
                                    Let $D$ be some dataset and let $\Theta = ${$\theta_1, \theta_2, \dotso, \theta_b$} be a set of parameters for a model. The goal of statistical inference
                                    is to find values for each parameter in $\Theta$ such that the model adequately describes the data $D$. <br><br>
                                    
                                    
                                    In Bayesian statistics, parameters are estimated using the posterior distribution. The posterior probability of $\Theta = \Theta_k$ can be obtained using:
                                    
                                    
                                    $$P(\Theta_k|D) = \frac{P(D|\Theta_k)P(\Theta_k)}{P(D)}$$
                                    
                                    where $P(D|\Theta_k)$ is the likelihood - the probability of observing data $D$ under the model assuming that the parameters in $\Theta_k$
                                    are set to their true values - and $P(\Theta_k)$ is the prior probability of $\Theta_k$ being the true values before observing new data $D$. $P(D)$ is a 
                                    normalisation constant and is intractable to evaluate directly. Subsequently, sampling methods are typically used to estimate the posterior distribution $P(\Theta|D)$.
                                    In a Bayesian framework, parameters are estimated as probability distributions
                                    - ie. the posterior distribution - instead of as point estimates. <br><br>
                                    
                                    
                                    Unfortunately, in some circumstances calculating the likelihood $P(D|\Theta)$ is not tractable. Instead,
                                    <a href="#Simulating" onclick="minimiseSection('#Simulating_Btn', true)" style="text-decoration:none; color:#008cba">simulation</a>                           
                                    can be used, and the posterior distribution can <i>approximated</i>. This framework is known as approximate Bayesian computation (ABC)<cite val="csillery2010approximate,beaumont2010approximate"></cite>.
                                    Let $\rho(D,\hat{D}_k)$ be a function which evaluates the distance between the true data $D$ and the data $\hat{D}_k$ simulated under
                                    the model with parameters $\Theta_k$. Then, $\Theta = \Theta_k$ is accepted into the posterior distribution if $\rho(D,\hat{D}_k) \leq \epsilon$
                                    for some threshold $\epsilon \geq 0$. The posterior distribution approximated under $\epsilon$ is denoted by $\hat{P}_\epsilon(\Theta_k|D)$. <br><br>
                                    
                                    
                                    If $\epsilon$ is sufficiently small, then $\hat{P}_\epsilon(\Theta_k|D)$ will be an accurate approximation of the true posterior distribution.
                                    However, if $\epsilon$ is too small, then adequately  sampling from $\hat{P}_\epsilon(\Theta_k|D)$ may not be attainable in any reasonable timeframe<cite val="bertorelle2010abc"></cite>. <br><br>
                                    
                                 
                                    In this section, first the various types of 
                                        <a href="#Experimental_Data" onclick="minimiseSection('#Approximate_Bayesian_computation_Btn', true)" style="text-decoration:none; color:#008cba">Experimental Data</a>
                                    $D$ supported by SimPol are described. Then, two ABC algorithms for
                                    estimating $\hat{P}_\epsilon(\Theta_k|D)$ are explained:
                                    
                                        <a href="#R-ABC" onclick="minimiseSection('#Approximate_Bayesian_computation_Btn', true)" style="text-decoration:none; color:#008cba">R-ABC</a>
                                     and
                                        <a href="#MCMC-ABC" onclick="minimiseSection('#Approximate_Bayesian_computation_Btn', true)" style="text-decoration:none; color:#008cba">MCMC-ABC</a>.
                                   
                                    
                                    
							</div>
						</div>
                            
                            
                            
                             
                            
                            
                            <div class="about-block">
                                <div class="about-block-inner">
                                <div class="subsection block-subtitle">Experimental Data</div> <br>
                                    
                                    
                                    
                                    <img src="../../src/TutorialImages/experimentalData.png" style="width:100%">
                                    
                                    
                                    
                                    SimPol supports three types of data, detailed below. The total experimental data is comprised of a series of $c$ experiments:
                                    $D = (d_1, d_2, \dotso, d_c)$. In the image above, there are $c=2$ experiments.
                                    Each datatype is associated with a distance function $\rho$. 
                                   The values of $\rho$ are summed across each of the $c$ experiments to give the total distance.
                                    
                                    
                                    
                                    
                                    
                                    <br><br>
                                    <div class="panelOuter">
                                        <div class="panelInner" >
                                            <span class="block-subsubtitle subsubsection" sect="Force-velocity" > Force-velocity data </span><br><br>
                                        
                                            In optical trapping experiments, varying forces are applied to the RNA polymerase using optical  
                                            tweezers<cite val="neuman2003ubiquitous,abbondanzieri2005direct,mejia2015trigger,dulin2015elongation"></cite>. As the assisting force increases, the rate of forward translocation
                                            increases with it.  <br><br>
                                            
                                            
                                            
                                             Experimental dataset $d_i \in D$ consists of a set of experimental settings $x$ and observations $y$. 
                                             
                                            $$d_i = \begin{bmatrix}
                                                        x_i^1 & y_i^1 \\
                                                        x_i^2 & y_i^2 \\
                                                        x_i^3 & y_i^3 \\
                                                        \vdots & \vdots
                                                    \end{bmatrix}$$
                                            
                                            
                                            
                                            
                                            Vector $x$ describes the varying values of the applied
                                                <a href="#Force" onclick="minimiseSection('#Parameters_and_Settings_Btn', true)" style="text-decoration:none; color:#008cba">Force</a>
                                            used in the experiment. The NTP concentrations are held constant at specified values. 
                                            The observations $y$ are the velocities of the polymerase, as measured under the applied force.   <br><br>
                                            
                                            
                                            For force-velocity data, $\rho$ is the chi-squared statistic $X^2$:
                                            
                                            
                                            $$
                                                \rho(d_i, \hat{d_i}) = X^2 = \sum_j \frac{(y_i^j - \hat{y}_j^j)^2}{\hat{y}_j^j}
                                            $$
                                            
                                            where $y_i$ are the true velocities for experiment $i$, and $\hat{y}_i$ are the simulated velocities.
                                            

                                    </div> 
                                    </div>  
                                    <br><br>
                                    
                                    
                                    
                                    
                                    <div class="panelOuter">
                                        <div class="panelInner" >
                                            <span class="block-subsubtitle subsubsection" sect="[NTP]-velocity" > [NTP]-velocity data </span><br><br>
                                        
                                            
                                            By increasing the concentrations of the four NTPs, the elongation velocity is expected to increase with it. <br><br>
                                            
                                            
                                            Similar to force-velocity data, experimental dataset $d_i \in D$ is expressed as a matrix with experimental settings vector $x$ and measurements $y$.
                                            
                                            For this data type, $x$ describes the varying values of the
                                                <a href="#[NTP]" onclick="minimiseSection('#Parameters_and_Settings_Btn', true)" style="text-decoration:none; color:#008cba">NTP concentrations</a>
                                            used in the experiment. Each NTP is associated with a concentration multiplier [ATP]<sub>eq</sub>, [CTP]<sub>eq</sub>, [GTP]<sub>eq</sub>, and [UTP]<sub>eq</sub>.
                                            The resulting concentrations for species _TP are computed as $x$[_TP]<sub>eq</sub>. The applied force is held constant at a specified value.
                                            The observations $y$ are the velocities of the polymerase, as measured under the NTP concentrations. <br><br>
                                            
                                            For [NTP]-velocity data, $\rho$ is the chi-squared statistic $X^2$, calculated the same as for force-velocity data.
                                            
                                             
                                    </div> 
                                    </div>  
                                    <br><br>
                                    
                                    
                                    
                                    <div class="panelOuter">
                                        <div class="panelInner" >
                                            <span class="block-subsubtitle subsubsection" sect="Pause sites" > Pause site data </span><br><br>
                                        
                                            At a pause site, the polymerase takes significantly longer than average to incorporate the next nucleotide and move on to the next position. <br><br>
                                            
                                            
                                            When specifying pause site data, the indicies of all pause sites (defined as the length of the nascent strand when the pause occurs) must be indicated, and so must the full nascent sequence.
                                            
                                            
                                            
                                            <br><br>
                                            Let $f_P(j)$ describe the median time that nascent strand has length $j$, averaged across multiple simulations of the specified gene.  
                                            SimPol treats pause site prediction as a binary classification problem. Subsequently, $\rho$ is equal to 1 minus the area under the curve (AUC) of a ROC curve <cite val="fawcett2006introduction"></cite>,
                                            computed by comparing the known locations of pause sites with continuous function $f_P$. A single AUC is computed across all pause site experimental datasets, not one per dataset.
                                            
                                             
                                    </div> 
                                    </div>  
                                    
                            </div>
                        </div>
                            
							
							
							
							
                            
                        
                        
                        <br><br>
                        <div class="about-block">
                            <div class="about-block-inner">
                            <div class="subsection block-subtitle">Motivating example 1</div> <br>
                                
                                
                                
                                <table style="width:100%">
                                        
                                      <tr>
                                           <td style="width:50%">
                                                <img src="../../src/TutorialImages/ABCexample1.png" style="width:100%">
                                                
                                           </td>
                                            
                                            
                                            <td style="width:50%">
                                                We will consider a simple example of using ABC to infer parameters of transcription elongation using a toy [NTP]-velocity dataset.
                                                The model assumes that NTP binding and translocation are at equilibrium and <param value="DGtau1"/> = 0.
                                                <param value="kCat"/> and <param value="Kdiss"/> are to be estimated and they have been assigned lognormal prior distributions. We will revisit this example
                                                using the R-ABC and MCMC-ABC algorithms.<br><br>
                                                
                                                
                                                   
                                                The posterior distribution can be summarised into a single state using
                                                <input type="button" value="Posterior Summary" title="Extract values from the posterior distribution" class="operation" style="background-color:#008cba; cursor:auto; width: 170px">.
                                                
                                                
                                                <br><br>
                                                
                                                
                                                       
                                                <br><br>For examples of ABC applied to real data, see 
                                                    <a href="#SimPol_in_the_Literature" onclick="minimiseSection('#SimPol_in_the_Literature_Btn', true)" style="text-decoration:none; color:#008cba">SimPol in the Literature</a>.
                                                    
                                           </td>
                                            
                                            
                                      </tr>
                                        
                                        
                                </table>
                                    
                                
                             
                                
                                
                            </div>
                        </div>
                            
                            
						
						
						<br><br>
						<div class="about-block">
							<div class="about-block-inner">
							<div class="subsection block-subtitle">R-ABC</div> <br>
								
								
								The rejection-ABC (R-ABC) algorithm is simple and easy to parallelise<cite val="csillery2010approximate,beaumont2010approximate"></cite>. Each sample is generated independently. The following user-specified settings are involved: 
								
								    <ul>
                                        <li><i>N iterations</i>, $N$: the total number of samples from posterior distribution $\hat{P}_\epsilon(\Theta|D)$ to generate. </li>
                                        <li><i>N trials per data point</i>, $n$: the number of simulations per data point. The average measured value (mean velocity or median pause time) will be taken. </li>
                                        <li>$\epsilon$: the threshold that $\rho(D,\hat{D}_k)$ must meet in order for $\Theta_k$ to be accepted into $\hat{P}_\epsilon(\Theta|D)$. This setting is linked with and can be used instead of $P(\rho \leq \epsilon)$.  </li>
                                        <li>$P(\rho \leq \epsilon)$: the probability that a value sampled from $P(\Theta)$ will be accepted. This setting is linked with and can be used instead of $\epsilon$. </li>
                                       
                                    </ul>
								
                                    
                                    
								<div style="padding: 2 10; background-color:#ebe9e7; font-size:18x" >
								
									<ol class="code">
									
									
										<li>
											function R_ABC($D$, $\Theta$, $N$, $\epsilon$) {
										</li>
									
										<li>
											&nbsp;&nbsp;&nbsp; $\hat{P}_\epsilon(\Theta|D) \leftarrow ${}
													<span class="comment"> // Initialise empty posterior distribution </span>
										</li>
										
										<li>
											&nbsp;&nbsp;&nbsp; for k = 1:N {
										</li>
										
										<li>
											&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\Theta_k \sim P(\Theta)$
														<span class="comment"> // Sample $\Theta_k$ from prior distribution </span>
										</li>
										
										
										
										
										<li>
											&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\hat{D}_k \sim $ simulator($\Theta_k$)
                                                        <span class="comment"> // Simulate data using the parameters specified in $\Theta_k$ </span>
										</li>
										
										
										<li>
											&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; if ($\rho(D, \hat{D}_k) \leq \epsilon$) {
										</li>
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\hat{P}_\epsilon(\Theta|D) \leftarrow \hat{P}_\epsilon(\Theta|D) \cup \Theta_k$
                                                <span class="comment"> // Accept </span>
                                        </li>
                                            
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; }
                                        </li>
                                            
                                            
                                            
										<li>
											&nbsp;&nbsp;&nbsp; }
										</li>
										
										
										<li>
											&nbsp;&nbsp;&nbsp; return $\hat{P}_\epsilon(\Theta|D)$
										</li>
										
										
										
										<li>
											}
										</li>
										
										
									</ol>
								
								</div>
                                    
                                    
                                <br>
                                To test this algorithm on motivating example 1, follow the link below and then find and press 
                                    
                                    <input type="button" value="Begin ABC" title="Begin the approximate Bayesian computation via rejection with the current settings" style="width: 170px; cursor: auto; background-color: rgb(0, 140, 186);" class="ABC_display operation beginABC_btn">
                                 <br><br>  
                                <span class="xml_link" style="text-align:center">
                                    <b>R-ABC motivating example 1 session:</b> 
                                    <a href="../?s=RABC.xml">Open</a> | 
                                    <a href="Examples/RABC.xml">View XML</a>
                                </span> 
                                    
                                    
	
							</div>
						</div>
                            
                            
                            
                            
                            
                            
                            
                            
                        <br><br>
                        <div class="about-block">
                            <div class="about-block-inner">
                            <div class="subsection block-subtitle">MCMC-ABC</div> <br>
                                
    
                                The Markov chain Monte Carlo ABC (MCMC-ABC) algorithm produces a random walk through parameter space<cite val="beaumont2010approximate"></cite>. Samples are no longer independent and therefore autocorrelation between successive states should be accounted for. 
                                In contrast to R-ABC, $\epsilon$ is not fixed. To enable $\rho$ to converge, $\epsilon$ starts out high at $\epsilon_0$ and then decreases to $\epsilon_{min}$ following
                                an exponential cooling scheme<cite val="ratmann2007using"></cite>: $\epsilon_i = \max(\epsilon_{min}, \gamma \epsilon_{i-1})$, where $0$ &lt; $\gamma$ &lt; $1$. 
                                $\epsilon_i$ refers to the value of the threshold at iteration $i$ in the MCMC chain.
                                If $\gamma$ is too small, the system may not converge, whereas if $\gamma$ is too large then it may take too long to converge.
                                
                                    
                                    The following user-specified settings are involved: 
                                    
                                    <ul>
                                        <li><i>N iterations</i>, $N$: the length of the chain of posterior distribution samples $\hat{P}_\epsilon(\Theta|D)$. </li>
                                        <li><i>N trials per data point</i>, $n$: the number of simulations per data point. The average measured value (mean velocity or median pause time) will be taken. </li>
                                        <li>$\epsilon_{min}$: the final value of $\epsilon$ after cooling.  </li>
                                        <li>$\epsilon_{0}$: the initial value of $\epsilon$. </li>
                                        <li>$\gamma$: the rate of $\epsilon$ cooling. </li>
                                        <li><i>Burn-in (%)</i>: the percentage of samples to discard before reaching the posterior distribution. Set to -1 to allow automatic detection. </li>
                                        <li><i>Log every (states)</i>, $L$: the frequency of storing posterior samples. Storing every sample is wasteful because of autocorrelation between successive sample. </li>
                                       
                                       
                                    </ul>
                                
                                    
                                    
                                <div style="padding: 2 10; background-color:#ebe9e7; font-size:18x" >
                                
                                    <ol class="code">
                                    
                                    
                                        <li>
                                            function MCMC_ABC($D$, $\Theta$, $N$, $\epsilon_0$, $\epsilon_{min}$, $\gamma$, $L$) {
                                        </li>
                                    
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; $\Theta_0 \sim P(\Theta)$
                                                    <span class="comment"> // Initial state sampled from prior distribution </span>
                                        </li>
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; $\hat{P}_\epsilon(\Theta|D) \leftarrow ${$\Theta_0$}
                                                    <span class="comment"> // Initialise posterior distribution </span>
                                        </li>
                                            
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; $\epsilon \leftarrow \epsilon_0$
                                                    <span class="comment"> // Initialise the value of $\epsilon$ </span>
                                        </li>
                                            
                                        
                                        <li>
                                            &nbsp;&nbsp;&nbsp; for k = 1:N {
                                        </li>
                                        
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\Theta^\prime \sim $ proposal$(\Theta_{k-1})$
                                                        <span class="comment"> // Sample proposal state using the proposal function. This randomly changes one parameter in $\Theta_{k-1}$. </span>
                                        </li>
                                        
                                        
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\hat{D}^\prime \sim $ simulator($\Theta^\prime$)
                                                        <span class="comment"> // Simulate data using the parameters specified in $\Theta^\prime$ </span>
                                        </li>
                                            
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\epsilon \leftarrow \max(\epsilon_{min}, \gamma \epsilon)$
                                                        <span class="comment"> // Lower the threshold </span>
                                        </li>
                                            
                                            
                                         <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $u \leftarrow$ random_uniform(0,1)
                                                        <span class="comment"> // A random number uniformly sampled from the range (0,1)</span>
                                        </li>
                                        
                                        
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; if ($\rho(D, \hat{D}^\prime) \leq \epsilon$ and $u$ &gt; $\frac{P(\Theta^\prime)}{P(\Theta_{k-1})}$) {
                                                        <span class="comment"> // Metropolis ratio<cite val="hastings1970monte"></cite> </span>
                                        </li>        
                                            
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\Theta_k \leftarrow \Theta^\prime$
                                                <span class="comment"> // Accept </span>
                                        </li>
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; } else {
                                                        
                                        </li>   
                                            
                                            
                                            
                                       <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\Theta_k \leftarrow \Theta_{k-1}$
                                                <span class="comment"> // Reject </span>
                                        </li
                                            
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; } 
                                                        
                                        </li>  
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; if ($k$ % $L$ = 0) {
                                        </li>
                                            
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; $\hat{P}_\epsilon(\Theta|D) \leftarrow \hat{P}_\epsilon(\Theta|D) \cup \Theta_k$
                                                <span class="comment"> // Log every $L$<sub>th</sub> state </span>
                                        </li>
                                        
                                          <li>
                                            &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; } 
                                        </li>  
                                            
                                        <li>
                                            &nbsp;&nbsp;&nbsp; }
                                        </li>
                                        
                                        
                                        <li>
                                            &nbsp;&nbsp;&nbsp; return $\hat{P}_\epsilon(\Theta|D)$
                                        </li>
                                        
                                        
                                        
                                        <li>
                                            }
                                        </li>
                                        
                                        
                                    </ol>
                                
                                </div>
                                    
                                    <br>
                                    To test this algorithm on motivating example 2, follow the link below and then find and press 
                                        
                                        <input type="button" value="Begin MCMC-ABC" title="Begin the approximate Bayesian computation via Markov chain Monte Carlo with the current settings" style="width: 170px; cursor: auto; background-color: rgb(0, 140, 186);" class="MCMC_display operation beginABC_btn">.
                                    
                                    The chain is not guaranteed to converge and may tay a long time even if it does.
                                     <br><br>  
                                    <span class="xml_link" style="text-align:center">
                                        <b>MCMC-ABC motivating example 2 session:</b> 
                                        <a href="../?s=MCMCABC.xml">Open</a> | 
                                        <a href="Examples/MCMCABC.xml">View XML</a>
                                    </span> 
             
                                    
                                    
    
                            </div>
                        </div>
                                
                           
							
							
						</div>
					</div>
				</div>
                
                
                
                
                
                
                <br><br><br><br>
                <div class="main-block">
                    <div class="main-block-inner">
                        <div class="block-title"> 
                            <span class="section">Model comparison</span> 
                        </div> 
                        <div style="display:none">
                
                            <div class="about-block">
                                <div class="about-block-inner">
    
                                    
                                    
                                    
                                    Recommended only for advanced users: the model itself can be estimated in 
                                        <a href="#Approximate_Bayesian_computation" onclick="minimiseSection('#Approximate_Bayesian_computation_Btn', true)" style="text-decoration:none; color:#008cba">Approximate Bayesian computation</a>
                                    as well as the various
                                        <a href="#Parameters_and_Settings" onclick="minimiseSection('#Parameters_and_Settings_Btn', true)" style="text-decoration:none; color:#008cba">Parameters</a>.
                                    Kinetic models, which may differ in their model settings (for example equilibrium assumptions, whether backtracking is enabled)
                                    or by constraining parameters at fixed values, can be specified in the Model comparison panel. Prior probabilities may be assigned to each
                                    model by giving each model a weight. The probability that the model will be activated at the beginning of each simulation,
                                    or the prior probability of that model during approximate Bayesian computation, is proportional to that weight. <br><br>
                                    
                                    Upon pressing 
                                        <input type="button" value="+ New model" title="Create a new model from the current settings" class="operation ABCbtn" style="background-color:#008CBA; width: 150px">
                                    a new kinetic model will be created. The settings of that model will be those that are currently specified. Any parameters which do not have
                                    a prior distribution will be held fixed at their current value, and those which do have a prior distribution will be omitted from the model description.
                                    The model description can be manually edited. <br><br>
                                    
                                      
                                    <img src="../../src/TutorialImages/modelComparison.png" style="width:100%; margin-top:20px; margin-bottom:20px">
                                    
                                    
                                     Refer to 
                                        <a href="#SimPol_in_the_Literature" onclick="minimiseSection('#SimPol_in_the_Literature_Btn', true)" style="text-decoration:none; color:#008cba">SimPol in the Literature</a>
                                    for real data examples of how model comparison has been used in inference.
                                    
                                    
                                    
                                </div>
                            </div>
                
                
                        </div>
                    </div>
                </div>

    
                
                
                
				
				<br><br><br><br>
				<div class="main-block">
					<div class="main-block-inner">
						<div class="block-title"> 
						
						
							<span class="section">Software Information</span> 
						</div> 
						<div style="display:none">
				
				
							<div style="display:none" class="about-block">
								<div class="about-block-inner">
									<div class="block-subtitle" > Web browser performance </div> <br>
										
										Rendering elements onto the screen (for example polymerase motion, plots, simulated sequences) can be computationally costly and can slow down the whole program.  If you are running a complex simulation on a 
										long sequence over a large number of trials, you may notice the simulation begin to slow down eventually. Maximising the performance of this program requires you as the user to regulate when rendering does or 
										does not occur. If the program starts to slow down, you should hide components of the interface with its respective minimise <span class='minimise' value='-' style="cursor:auto; padding: 1 1">&minus;</span>
										button. If a component is minimised then rendering will not occur. The polymerase animation can be minimised by setting the speed to <b>hidden</b>. This will happen automatically when simulating large sequences. <br><br>
										The simulation component of SimPol is written in C++ and compiled with <a href="http://webassembly.org/">WebAssembly</a>. You will be notified if your web browser fails to instantiate the WebAssembly module.
										In this case, the simulation will instead be run using the JavaScript module which will have a significant impact on performance.    <br><br>
										
										
										In January 2018 we tested SimPol on different operating systems using recent versions of different web browsers, using both JavaScript (JS) and C++. 
										Each setting was tested on 1000 simulations of the 4029nt E.coli ropB gene (accession number KP670789). 
										Compiling the C++ code with WebAssembly increases speed performance approximately 10x over the same algorithm written in JavaScript for the fastest JavaScript environments. 
										For maximum performance we recommend using SimPol on Firefox, Chrome, Safari, or through the command line.
										
										
										<span style="padding: 1 5; background-color:#ebe9e7"> 
											<b style="color:#008cba">Benchmark session:</b> 
											<a style="text-decoration:none" href="../?s=benchmark.xml">Open</a> | 
											<a style="text-decoration:none" href="Examples/benchmark.xml">View XML</a>
										</span>										
										<br> <br>
										
										<table cellpadding=5 style="margin:auto; text-align:center;">
											<tr style="background-color:#b3b3b3; font-size:18px">
												<td>
													OS
												</td>
												<td>
													Device description
												</td>
												<td>
													Program (SimPol language)
												</td>
												<td>
													Time (min)
												</td>
											</tr>
											
											
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td rowspan=6  style="background-color:#E95420">
													Ubuntu 7
												</td>
												
												<td rowspan=6>
													Intel i5 3.30GHz
												</td>
	
												<td>
													<b>Command prompt (C++)</b>
												</td>
												
												<td>
													
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
									
	
												<td>
													<b>Node.js 8.6 (JS)</b>
												</td>
												
												<td>
													3:02
												</td>
												
											</tr>
											<tr style="background-color:#E95420; color:white; font-size:18px">
												
												
												<td >
													<b>Chrome 61 (C++)</b>
												</td>
												
												<td>
													0:15
												</td>
												
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												
												<td >
													<b>Chrome 61 (JS)</b>
												</td>
												
												<td>
													3:22
												</td>
												
											</tr>
											
											
											<tr style="background-color:#E95420; color:white; font-size:18px">
												
											
												<td >
													<b>Firefox 58 (C++)</b>
												</td>
												
												<td>
													0:14
												</td>
												
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
											
												<td >
													<b>Firefox 58 (JS)</b>
												</td>
												
												<td>
													5:14
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td rowspan=9  style="background-color:#00a1f1">
													Windows 10
												</td>
												
												<td rowspan=9>
													AMD Ryzen 5 1600 3.20 GHz
												</td>
												
											
												<td>
													<b>Command prompt (C++)</b>
												</td>
												
												<td>
													
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
											
												
											
												<td>
													<b>Cygwin (C++)</b>
												</td>
												
												<td>
													0:37
												</td>
												
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
											
												
											
												<td>
													<b>Node.js 8.7 (JS)</b>
												</td>
												
												<td>
													2:49
												</td>
												
											</tr>
											
											
											
											
											<tr style="background-color:#00a1f1; color:white; font-size:18px">
												
											
											
												<td >
													<b>Chrome 63 (C++)</b>
												</td>
												
												<td>
													0:15
												</td>
												
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
											
											
												<td >
													<b>Chrome 63 (JS)</b>
												</td>
												
												<td>
													2:37
												</td>
												
											</tr>
											
											
																						
											<tr style="background-color:#00a1f1; color:white; font-size:18px">
												
											
											
												<td >
													<b>Firefox 58 (C++)</b>
												</td>
												
												<td>
													0:16
												</td>
												
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
											
											
												<td >
													<b>Firefox 58 (JS)</b>
												</td>
												
												<td>
													4:37
												</td>
												
											</tr>
											
											
											
																						
											<tr style="background-color:#ebe9e7; font-size:18px">
												
											
											
												<td >
													<b>Microsoft Edge 40 (C++)</b>
												</td>
												
												<td>
													
												</td>
												
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
											
											
												<td >
													<b>Microsoft Edge 40 (JS)</b>
												</td>
												
												<td>
													15:33
												</td>
												
											</tr>
											
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												<td rowspan=3  style="background-color:#999999">
													OS X 10.13
												</td>
												<td rowspan=3>
													Intel i5 3.0GHz
												</td>
												<td style="background-color:#999999; color:white; font-size:18px">
													<b>Command prompt (C++)</b>
												</td>
												<td style="background-color:#999999; color:white; font-size:18px">
													0:14
												</td>
											</tr>
											
											
											<tr style="background-color:#999999; color:white; font-size:18px">
											
												<td>
													<b>Safari 10.3 (C++)</b>
												</td>
												<td>
													0:13
												</td>
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
											
											
												<td>
													<b>Safari 10.3 (JS)</b>
												</td>
												
												<td>
													2:38
												</td>
												
											</tr>
		
											
											
										
											
											
												
											<tr style="background-color:#ebe9e7; font-size:18px">
												<td rowspan=2  style="background-color:#999999">
													iOS 10.0.3
												</td>
												
												<td rowspan=2>
													iPhone 7
												</td>
												<td >
													<b>Chrome 61 (JS)</b>
												</td>
												
												
												<td>
													
												</td>
												
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
		
												
												<td >
													<b>Safari 10.3 (JS)</b>
												</td>
												
												
												<td>
													
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td rowspan=4  style="background-color:#a4c639">
													Android 7.0
												</td>
												
												<td rowspan=4>
													Samsung Galaxy S6 Edge+
												</td>
												
											
												<td style="background-color:#a4c639; color:white">
													<b>Chrome 64 (C++)</b>
												</td>
												
												<td style="background-color:#a4c639; color:white">
													0:52
												</td>
												
											</tr>
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												
											
												<td >
													<b>Chrome 64 (JS)</b>
												</td>
												
												<td>
													8:03
												</td>
												
											</tr>
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												
											
												<td style="background-color:#a4c639; color:white">
													<b>Firefox 58 (C++)</b>
												</td>
												
												<td style="background-color:#a4c639; color:white">
													0:46
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												
											
												<td >
													<b>Firefox 58 (JS)</b>
												</td>
												
												<td>
													11:23
												</td>
												
											</tr>
											
											
										
								</table>
	
								</div>
							</div>
							<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									<div id="fromCommandLine_performance" class="subsection block-subtitle"> Running SimPol from the command line </div> <br>
	
                                    
                                        The following instructions will work for Linux/Mac. 
                                    
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b> 1. Generate a SimPol session file.</b> Go to <a href="http://www.polymerase.nz">SimPol</a> in your web browser and configure the number of trials, parameters, model settings, ABC settings, etc.
											When you are happy with all the settings, download the session as an XML file <img id='saveSessionID'  src="../../src/Images/downloadGreen.png" style="padding:5 0; vertical-align:middle; height:25px;">.
										</div> <br>
										
										
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b> 2. Download SimPol </b> from <a href="https://github.com/jordandouglas/jordandouglas.github.io">here</a> and unzip. Open up the command terminal and navigate to this folder.
										</div> <br>
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b> 3. Compile the ViennaRNA suite module.</b> RNAfold<cite val="lorenz2011viennarna"></cite> is used when <param value="allowmRNAfolding"/> is enabled. <br><br>
											
											
											<div style="margin-left: 100px;">
													
												<span class="code">cd src/asm/ViennaRNA</span><br>
												<span class="code">bash build_vrna.sh </span><br>
											</div>
										</div> <br>
                                    
                                    
                                    
                                        <div style="padding: 10 10; background-color:#ebe9e7" >
                                                <b> 4. Compile SimPol. </b> ViennaRNA must be successfully compiled first.<br><br>
                                            
                                            
                                            <div style="margin-left: 100px;">
                                                    
                                                <span class="code">cd ../../../simpol/src/asm/</span><br>
                                                <span class="code">bash build_vrna.sh </span><br>
                                            </div>
                                        </div> <br>
										
										<div style="padding: 10 10; background-color:#ebe9e7" >
											<b> 5. Run SimPol.</b> <br><br>
												<div class="code" style="margin-left: 100px;">
													
													bash simpol.sh -xml path/to/session.xml
												</div> <br>
												where <span class="code">session.xml</span> is the session generated earlier. This will run the number of simulations
											     specified in the .xml file. Optional arguments:<br><br>
												<div style="margin-left: 100px;">
													
											        <span class="code">-h</span>: View the command line argument specifications. <br><br>
                                                   		
                                            
													<span class="code">-logO file.log</span>: Write the MCMC-ABC or R-ABC output into the specified file. <br><br>
                                                   
													<span class="code">-nthreads N</span>: Parallelise the simulations over N threads. Multithreading does not work in the web browser, only from the command line. Default: 1. <br><br>
                                            
                                                    <span class="code">-MCMC</span>: Run MCMC-ABC analysis instead of simulating. <br><br>
                                                    <span class="code">-resume</span>: Resumes MCMC from the last state printed in the file specified by <span class="code">-logO</span> and appends to <span class="code">-logO</span>. Only applicable if <span class="code">-MCMC</span> is enabled. <br><br>
                                                    
                                            
                                                    <span class="code">-logI file.log</span>: Read in the specified file. Only applicable if either -summary or -RABC are activated. <br><br>
                                                    <span class="code">-summary</span>: prints a summary of the log file specified by <span class="code">-logI</span> into the terminal. <br><br>
                                                    <span class="code">-RABC</span>: Run R-ABC analysis instead of simulating. All states will be printed regardless of the xml-specified value of epsilon. 
                                                                                    If <span class="code">-logI</span> is specified then will sample from the posterior distribution, otherwise the prior.
                                                                                    If <span class="code">-summary</span> is also specified then will sample from the geometric median instead of the whole posterior distribution. <br><br>
                                                    
                                                  
                                            
												</div>
										</div>
								</div>
							</div>
				
				
						</div>
					</div>
				</div>

	
	<!--
	<br><br><br><br>
	<div class="main-block">
		<div class="main-block-inner">
			<div class="block-title"> 
			
			
				Kinetic Data 
				</div> 
					<div id="data"  style="display:none">
						<br><br>
							<div class="about-block">
								<div class="about-block-inner">
									
									Transcription elongation velocities (bp/s) below. <br><br>
									
									
									
										<table cellpadding=5 style="margin:auto; text-align:center; vertical-align:middle">
											<tr style="font-size:18px">
												<td>
													
												</td>
												<td  style="background-color:#b3b3b3;">
													RNA polymerase
												</td>
												<td style="background-color:#b3b3b3;">
													Optical trapping
												</td>
												<td style="background-color:#b3b3b3;">
													Other technique
												</td>
											</tr>
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												<td rowspan=1 style="background-color:#b3b3b3;">
													Prokaryote
												</td>
												<td>
													<i>Escherichia coli</i> RNAP
												</td>
												
												<td>
													7.3 bp/s <cite val="davenport2000single"></cite> <br>
													9.4 bp/s <cite val="neuman2003ubiquitous"></cite> <br>
													11.7 bp/s <cite val="mejia2015trigger"></cite> <br>
													12.8 bp/s <cite val="adelman2002single"></cite> <br>
													15.9 bp/s <cite val="tolic2004diversity"></cite> <br>
													20.6 bp/s <cite val="bai2007mechanochemical"></cite> <br>
													23 bp/s <cite val="abbondanzieri2005direct"></cite>
												</td>
												
												<td>
													25 bp/s <cite val="iost1992bacteriophage"></cite> <br>
													28-89 bp/s <cite val="vogel1994rna"></cite> <br>
													30-103 bp/s <cite val="ryals1982temperature"></cite> <br>
													39-85 bp/s <cite val="bremer1996modulation"></cite>  <br>
													
													90 bp/s <cite val="richardson1996control"></cite> <br>
													
													
													
													
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												<td rowspan=5 style="background-color:#b3b3b3;">
													Eukaryote
												</td>
												<td>
													<i>Human</i> pol II
												</td>
												
												<td>
													
													
												</td>
												
												<td>
												
												
													<table style="width:100%; text-align:center;  font-size:18px;" cellpadding=5>
												
														<tr>
														
															<td>
																<i> FRAP </i> <br>
																30 bp/s <cite val="boireau2007transcriptional"></cite> <br>
																72 bp/s <cite val="darzacq2007vivo"></cite> <br><br>
															</td>
															
															<td>
																<i> qRT-PCR </i> <br>
																40 bp/s <cite val="tennyson1995human"></cite> <br>
																63 bp/s <cite val="singh2009rates"></cite> <br><br>
															</td>
															
														</tr>
														
														
															<tr>
														
															<td>
																
																50-100 bp/s <cite val="sehgal19765"></cite> <br>
															</td>
															
															<td>
															</td>
															
														</tr>
													
													</table>
													
													
													
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td>
													<i>Saccharomyces cerevisiae</i> pol I
												</td>
												
												<td>
													
													
												</td>
												
												<td>
												
													40 bp/s <cite val="kovs2010yeast"></cite> <br>
													60 bp/s <cite val="french2003exponentially"></cite> <br>
													
													
													
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td>
													<i>Saccharomyces cerevisiae</i> pol II
												</td>
												
												<td>
													10-20 bp/s <cite val="galburt2009single"></cite> <br>
													17.5 bp/s <cite val="hodges2009nucleosomal"></cite> <br>
													25 bp/s <cite val="dangkulwanich2013complete"></cite> <br>
													
													
													
												</td>
												
												<td>
												
													33.3 bp/s <cite val="mason2005distinction"></cite> <br>
													
													
													
													
												</td>
												
											</tr>
											
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td>
													<i>Rat</i> pol II
												</td>
												
												<td>
													
													
												</td>
												
												<td>
												
													20-25 bp/s <cite val="ucker1984early"></cite> <br>
													22 bp/s <cite val="mejia2015trigger"></cite> <br>
													
													
												</td>
												
											</tr>
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td>
													<i>Drosophila</i> pol II
												</td>
												
												<td>
													
													
												</td>
												
												<td>
												
													18 bp/s <cite val="thummel1990spatial"></cite> <br>
													22 bp/s <cite val="o1993rapid"></cite> <br>
													23 bp/s <cite val="shermoen1991progression"></cite> <br>
													25 bp/s <cite val="yao2007intranuclear"></cite> <br>
													
													
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												<td rowspan=1 style="background-color:#b3b3b3;">
													dsDNA virus
												</td>
												<td>
													<i>Bacteriophage T7</i> pol
												</td>
												
												<td>
													43 bp/s <cite val="skinner2004promoter"></cite> <br>
													98-141 bp/s <cite val="thomen2008t7"></cite> <br>
													
												</td>
												
												<td>
													200 bp/s <cite val="iost1992bacteriophage"></cite> <br>
													222 bp/s <cite val="anand2006transient"></cite> <br>
													240 bp/s <cite val="makarova1995transcribing"></cite> <br>
													
			
												</td>
												
											</tr>
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												<td rowspan=1 style="background-color:#b3b3b3;">
													dsRNA virus
												</td>
												<td>
													<i>Bacteriophage &Phi;6</i> P2
												</td>
												
												<td>
													16.7 bp/s <cite val="dulin2015elongation"></cite> <br>
													
												</td>
												
												<td>
													20-30 bp/s <cite val="kainov2004packaging"></cite> <br>
													
			
												</td>
												
											</tr>
											
											
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												<td rowspan=3 style="background-color:#b3b3b3;">
													(-)ssRNA virus
												</td>
												<td>
													<i>Sendai virus</i> L
												</td>
												
												<td>
													
													
												</td>
												
												<td>
													1.7 bp/s <cite val="gubbay2001sendai"></cite> <br>
													
			
												</td>
												
											</tr>
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td>
													<i>Measles virus</i> L
												</td>
												
												<td>
													
													
												</td>
												
												<td>
													3 bp/s <cite val="plumet2005dynamics"></cite> <br>
													
			
												</td>
												
											</tr>
											
											
											<tr style="background-color:#ebe9e7; font-size:18px">
												
												<td>
													<i>Vesicular stomatitis virus</i> L
												</td>
												
												<td>
													
													
												</td>
												
												<td>
													3.1-4.3 bp/s <cite val="iverson1981localized"></cite> <br>
													
			
												</td>
												
											</tr>
										
											
											
										
								</table>
								</div>
							</div>
					</div>
				</div>
	
	
			</div>
		</div>
	</div>
	-->
	
                    
                    
    <br><br><br><br>
    <div class="main-block">
        <div class="main-block-inner">
            <div class="block-title"> 
            
            
                    <span class="section">SimPol in the Literature </span>
                </div> 
                <div style="display:none">          
                    <br><br>
                    <div class="about-block" style="width:90%; margin:auto">
                        <div class="about-block-inner">
                                    
                                    <div class="subsection block-subtitle">Bayesian inference and comparison of stochastic transcription elongation models</div> <br> 
                                    
                                    In this project (publication under review), we inferred kinetic parameters for three RNA polymerases: <i>E. coli</i> RNAP,  
                                    <i>S. cerevisiae</i> pol II, and Bacteriophage T7 pol. The posterior distributions for these three experiments are available below.
                                    Subsampled posterior distributions may be viewed directly in the web browser, and the full posterior distributions are available to download.
                                    Downloaded .log files can be viewed with Tracer<cite val="rambaut2015tracer"></cite>.  <br><br>
                                    
                                    <table style="padding: 10; text-align:center;" cellpadding="5">
                                        
                                       <tr style="background-color:#b3b3b3; color:white;">
                                            <th style="width:200px">
                                                RNA polymerase
                                            </th>
                                            
                                            <th style="width:150px">
                                                View sample
                                            </th>
                                            
                                            <th style="width:150px">
                                                Download .xml
                                            </th>
                                            
                                            <th style="width:150px">
                                                Download .log
                                            </th>
                                            
                                       </tr>
                                        
                                        
                                        <tr>
                                            <td>
                                                <i>E. coli</i> RNAP
                                            </td>
                                            
                                            <td>
                                                <a href="../?ploscompbio=1">Open</a>
                                            </td>
                                            
                                            <td>
                                                <a href="ElongationPosteriorData/Ecoli_RNAP_session.xml">Ecoli_RNAP_session.xml</a>
                                            </td>
                                            
                                            <td>
                                                <a href="ElongationPosteriorData/Ecoli_RNAP_posterior.log">Ecoli_RNAP_posterior.log</a>
                                            </td>
                                            
                                       </tr>
                                        
                                        
                                        <tr>
                                            <td>
                                                <i>S. cerevisiae</i> pol II
                                            </td>
                                            
                                            <td>
                                                <a href="../?ploscompbio=2">Open</a>
                                            </td>
                                           
                                            
                                            <td>
                                                <a href="ElongationPosteriorData/Yeast_polII_session.xml">Yeast_polII_session.xml</a>
                                            </td>
                                            
                                            <td>
                                                <a href="ElongationPosteriorData/Yeast_polII_posterior.log">Yeast_polII_posterior.log</a>
                                            </td>
                                            
                                       </tr>
                                        
                                        
                                        <tr>
                                            <td>
                                                Bacteriophage T7 pol
                                            </td>
                                            
                                            <td>
                                                <a href="../?ploscompbio=3">Open</a>
                                            </td>
                                            
                                            
                                            <td>
                                                <a href="ElongationPosteriorData/T7_pol_session.xml">T7_pol_session.xml</a>
                                            </td>
                                            
                                            <td>
                                                <a href="ElongationPosteriorData/T7_pol_posterior.log">T7_pol_posterior.log</a>
                                            </td>
                                            
                                       </tr>
                                        
                                        
                                    </table>
                                    
                                    
                                    
                                    <br><br>
                                    <div class="ref">Douglas, Jordan, Richard Kingston, and Alexei Drummond. "Bayesian inference and comparison of stochastic transcription elongation models." bioRxiv (2018): 499277.
                                        <span style="background-color:#ebe9e7; padding: 2 4" title="https://doi.org/10.1101/499277 ">
                                            <b style="color:#008cba">DOI:</b>
                                            <a style="text-decoration:none; color:#008cba" href="https://doi.org/10.1101/499277 ">view</a>
                                        </span>
                                    </div>
  
                                    
                        </div>
                    </div>
                </div>
    
    
            </div>
        </div>
    </div>
                    
                    
                    
	
	<br><br><br><br>
	<div class="main-block">
		<div class="main-block-inner">
			<div class="block-title"> 
			
			
				    <span class="section">References</span>
				</div> 
					<div style="display:none">
						<br><br>
						<div class="about-block" style="width:90%; margin:auto">
							<div class="about-block-inner"  id="references">
									
									
									
							</div>
						</div>
					</div>
				</div>
	
	
			</div>
		</div>
	</div>
	
	
	
	
	
	<br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br>
	<br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> <br> 
	
	
		<div class="header">
			SimPol | Centre for Computational Evolution & School of Biological Sciences, University of Auckland | 2019
		</div>
	
		</div>
	</div>
</div>
</body>
</html>
